<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Work with other organisms • rGREAT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Work with other organisms">
<meta property="og:description" content="rGREAT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rGREAT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Online GREAT</li>
    <li>
      <a href="../articles/1.online-GREAT.html">Analyze with online GREAT</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Local GREAT</li>
    <li>
      <a href="../articles/2.local-GREAT.html">Analyze with local GREAT</a>
    </li>
    <li>
      <a href="../articles/3.other-geneset-databases.html">Work with other geneset databases</a>
    </li>
    <li>
      <a href="../articles/4.other-organisms.html">Work with other organisms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Supplementaries in the rGREAT paper</li>
    <li>
      <a href="../articles/suppl_use_background.html">Suppl 1. Use background regions</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_GO.html">Suppl 2. Compare GO annotations from online and local GREAT</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_online_and_local.html">Suppl 3. Compare online and local GREAT</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_tss.html">Suppl 4. Compare different TSS sources</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_genome_version.html">Suppl 5. Compare different genome versions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/rGREAT/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Work with other organisms</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu ( <a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a> )</h4>
            
            <h4 data-toc-skip class="date">2024-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/rGREAT/blob/HEAD/vignettes/4.other-organisms.Rmd" class="external-link"><code>vignettes/4.other-organisms.Rmd</code></a></small>
      <div class="hidden name"><code>4.other-organisms.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="use-txdb-packages">Use TxDb packages<a class="anchor" aria-label="anchor" href="#use-txdb-packages"></a>
</h2>
<p>In the Bioconductor annotation ecosystem, there are <strong>TxDb.*</strong> packages which provide data for Gene Ontology gene sets. The <strong>TxDb.*</strong> packages supported in <strong>rGREAT</strong> are:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/rGREAT" class="external-link">rGREAT</a></span><span class="op">)</span></span>
<span><span class="fu">rGREAT</span><span class="fu">:::</span><span class="va">BIOC_ANNO_PKGS</span><span class="op">$</span><span class="va">txdb</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "TxDb.Hsapiens.UCSC.hg18.knownGene"     </span></span>
<span><span class="co">##  [2] "TxDb.Hsapiens.UCSC.hg19.knownGene"     </span></span>
<span><span class="co">##  [3] "TxDb.Hsapiens.UCSC.hg38.knownGene"     </span></span>
<span><span class="co">##  [4] "TxDb.Hsapiens.UCSC.hg38.refGene"       </span></span>
<span><span class="co">##  [5] "TxDb.Mmusculus.UCSC.mm10.knownGene"    </span></span>
<span><span class="co">##  [6] "TxDb.Mmusculus.UCSC.mm10.ensGene"      </span></span>
<span><span class="co">##  [7] "TxDb.Mmusculus.UCSC.mm39.refGene"      </span></span>
<span><span class="co">##  [8] "TxDb.Mmusculus.UCSC.mm9.knownGene"     </span></span>
<span><span class="co">##  [9] "TxDb.Rnorvegicus.UCSC.rn4.ensGene"     </span></span>
<span><span class="co">## [10] "TxDb.Rnorvegicus.UCSC.rn5.refGene"     </span></span>
<span><span class="co">## [11] "TxDb.Rnorvegicus.UCSC.rn6.refGene"     </span></span>
<span><span class="co">## [12] "TxDb.Rnorvegicus.UCSC.rn7.refGene"     </span></span>
<span><span class="co">## [13] "TxDb.Ggallus.UCSC.galGal4.refGene"     </span></span>
<span><span class="co">## [14] "TxDb.Ggallus.UCSC.galGal5.refGene"     </span></span>
<span><span class="co">## [15] "TxDb.Ggallus.UCSC.galGal6.refGene"     </span></span>
<span><span class="co">## [16] "TxDb.Mmulatta.UCSC.rheMac10.refGene"   </span></span>
<span><span class="co">## [17] "TxDb.Mmulatta.UCSC.rheMac3.refGene"    </span></span>
<span><span class="co">## [18] "TxDb.Mmulatta.UCSC.rheMac8.refGene"    </span></span>
<span><span class="co">## [19] "TxDb.Celegans.UCSC.ce11.refGene"       </span></span>
<span><span class="co">## [20] "TxDb.Celegans.UCSC.ce11.ensGene"       </span></span>
<span><span class="co">## [21] "TxDb.Cfamiliaris.UCSC.canFam3.refGene" </span></span>
<span><span class="co">## [22] "TxDb.Cfamiliaris.UCSC.canFam4.refGene" </span></span>
<span><span class="co">## [23] "TxDb.Cfamiliaris.UCSC.canFam5.refGene" </span></span>
<span><span class="co">## [24] "TxDb.Sscrofa.UCSC.susScr11.refGene"    </span></span>
<span><span class="co">## [25] "TxDb.Sscrofa.UCSC.susScr3.refGene"     </span></span>
<span><span class="co">## [26] "TxDb.Scerevisiae.UCSC.sacCer2.sgdGene" </span></span>
<span><span class="co">## [27] "TxDb.Scerevisiae.UCSC.sacCer3.sgdGene" </span></span>
<span><span class="co">## [28] "TxDb.Ptroglodytes.UCSC.panTro4.refGene"</span></span>
<span><span class="co">## [29] "TxDb.Ptroglodytes.UCSC.panTro5.refGene"</span></span>
<span><span class="co">## [30] "TxDb.Ptroglodytes.UCSC.panTro6.refGene"</span></span>
<span><span class="co">## [31] "TxDb.Dmelanogaster.UCSC.dm3.ensGene"   </span></span>
<span><span class="co">## [32] "TxDb.Dmelanogaster.UCSC.dm6.ensGene"   </span></span>
<span><span class="co">## [33] "TxDb.Drerio.UCSC.danRer10.refGene"     </span></span>
<span><span class="co">## [34] "TxDb.Drerio.UCSC.danRer11.refGene"     </span></span>
<span><span class="co">## [35] "TxDb.Btaurus.UCSC.bosTau8.refGene"     </span></span>
<span><span class="co">## [36] "TxDb.Btaurus.UCSC.bosTau9.refGene"     </span></span>
<span><span class="co">## [37] "TxDb.Athaliana.BioMart.plantsmart51"   </span></span>
<span><span class="co">## [38] "TxDb.Athaliana.BioMart.plantsmart22"   </span></span>
<span><span class="co">## [39] "TxDb.Athaliana.BioMart.plantsmart25"   </span></span>
<span><span class="co">## [40] "TxDb.Athaliana.BioMart.plantsmart28"</span></span></code></pre>
<p>To perform GREAT anlaysis with GO gene sets for other organisms, you can either specify the genome version:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"galGal6"</span><span class="op">)</span></span></code></pre></div>
<p>or with the full name of the corresponding TxDb package:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"TxDb.Ggallus.UCSC.galGal6.refGene"</span><span class="op">)</span></span></code></pre></div>
<p>These two are internally the same.</p>
</div>
<div class="section level2">
<h2 id="use-biomart-go-gene-sets">Use BioMart GO gene sets<a class="anchor" aria-label="anchor" href="#use-biomart-go-gene-sets"></a>
</h2>
<p>You can specify a BioMart dataset (which corresponds to a specific organism), e.g.:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Giant panda</span></span>
<span><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, biomart_dataset <span class="op">=</span> <span class="st">"amelanoleuca_gene_ensembl"</span><span class="op">)</span></span></code></pre></div>
<p style="color:red">
Make sure the genome version fit your data.
</p>
<p>A full list of supported BioMart datasets (organisms) as well as the genomo versions can be found with the function <code><a href="https://rdrr.io/pkg/BioMartGOGeneSets/man/supportedOrganisms.html" class="external-link">BioMartGOGeneSets::supportedOrganisms()</a></code>.</p>
</div>
<div class="section level2">
<h2 id="use-msigdb-gene-sets">Use MSigDB gene sets<a class="anchor" aria-label="anchor" href="#use-msigdb-gene-sets"></a>
</h2>
<p>MSigDB contains gene sets only for human, but it can be extended to other organisms by mapping to the homologues genes. The package <a href="https://cran.r-project.org/web/packages/msigdbr/index.html" class="external-link"><strong>msigdbr</strong></a> has already mapped genes to many other organisms. A full list of supported organisms in <strong>msigdbr</strong> can be obtained by：</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://igordot.github.io/msigdbr/" class="external-link">msigdbr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr_species.html" class="external-link">msigdbr_species</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 20 × 2</span></span></span>
<span><span class="co">##    species_name                    species_common_name                          </span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                           <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> Anolis carolinensis             Carolina anole, green anole                  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> Bos taurus                      bovine, cattle, cow, dairy cow, domestic cat…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> Caenorhabditis elegans          <span style="color: #BB0000;">NA</span>                                           </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> Canis lupus familiaris          dog, dogs                                    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> Danio rerio                     leopard danio, zebra danio, zebra fish, zebr…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> Drosophila melanogaster         fruit fly                                    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> Equus caballus                  domestic horse, equine, horse                </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> Felis catus                     cat, cats, domestic cat                      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> Gallus gallus                   bantam, chicken, chickens, Gallus domesticus </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> Homo sapiens                    human                                        </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">11</span> Macaca mulatta                  rhesus macaque, rhesus macaques, Rhesus monk…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">12</span> Monodelphis domestica           gray short-tailed opossum                    </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">13</span> Mus musculus                    house mouse, mouse                           </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">14</span> Ornithorhynchus anatinus        duck-billed platypus, duckbill platypus, pla…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">15</span> Pan troglodytes                 chimpanzee                                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">16</span> Rattus norvegicus               brown rat, Norway rat, rat, rats             </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">17</span> Saccharomyces cerevisiae        baker's yeast, brewer's yeast, S. cerevisiae </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">18</span> Schizosaccharomyces pombe 972h- <span style="color: #BB0000;">NA</span>                                           </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">19</span> Sus scrofa                      pig, pigs, swine, wild boar                  </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">20</span> Xenopus tropicalis              tropical clawed frog, western clawed frog</span></span></code></pre>
<p>To obtain gene sets for non-human organisms, e.g.:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">h_gene_sets</span> <span class="op">=</span> <span class="fu"><a href="https://igordot.github.io/msigdbr/reference/msigdbr.html" class="external-link">msigdbr</a></span><span class="op">(</span>species <span class="op">=</span> <span class="st">"chimpanzee"</span>, category <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">h_gene_sets</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 6 × 18</span></span></span>
<span><span class="co">##   gs_cat gs_subcat gs_name               gene_symbol entrez_gene ensembl_gene   </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                 <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> H      <span style="color: #949494;">""</span>        HALLMARK_ADIPOGENESIS ABCA1            <span style="text-decoration: underline;">464</span>630 ENSPTRG0000002…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> H      <span style="color: #949494;">""</span>        HALLMARK_ADIPOGENESIS ABCB8            <span style="text-decoration: underline;">463</span>892 ENSPTRG0000001…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> H      <span style="color: #949494;">""</span>        HALLMARK_ADIPOGENESIS ACAA2            <span style="text-decoration: underline;">455</span>414 ENSPTRG0000001…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> H      <span style="color: #949494;">""</span>        HALLMARK_ADIPOGENESIS ACADL            <span style="text-decoration: underline;">459</span>914 ENSPTRG0000001…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> H      <span style="color: #949494;">""</span>        HALLMARK_ADIPOGENESIS ACADM            <span style="text-decoration: underline;">469</span>356 ENSPTRG0000000…</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> H      <span style="color: #949494;">""</span>        HALLMARK_ADIPOGENESIS ACADS            <span style="text-decoration: underline;">742</span>921 ENSPTRG0000000…</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 12 more variables: human_gene_symbol &lt;chr&gt;, human_entrez_gene &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   human_ensembl_gene &lt;chr&gt;, gs_id &lt;chr&gt;, gs_pmid &lt;chr&gt;, gs_geoid &lt;chr&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   gs_exact_source &lt;chr&gt;, gs_url &lt;chr&gt;, gs_description &lt;chr&gt;, taxon_id &lt;int&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   ortholog_sources &lt;chr&gt;, num_ortholog_sources &lt;dbl&gt;</span></span></span></code></pre>
<p>If the organism you selected has a corresponding TxDb package available which provides TSS information, you need to make sure the gene sets use Entrez gene ID as the identifier (Most TxDb packages use Entrez ID as primary ID, you can check the variable <code>rGREAT:::BIOC_ANNO_PKGS</code>).</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># convert to a list of gene sets</span></span>
<span><span class="va">h_gene_sets</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="va">h_gene_sets</span><span class="op">$</span><span class="va">entrez_gene</span>, <span class="va">h_gene_sets</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span></span>
<span><span class="va">h_gene_sets</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">h_gene_sets</span>, <span class="va">as.character</span><span class="op">)</span>  <span class="co"># just to make sure gene IDs are all in character.</span></span>
<span><span class="va">h_gene_sets</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## $HALLMARK_ADIPOGENESIS</span></span>
<span><span class="co">##   [1] "464630"    "463892"    "455414"    "459914"    "469356"    "742921"   </span></span>
<span><span class="co">##   [7] "454672"    "104003784" "454895"    "451866"    "737339"    "471032"   </span></span>
<span><span class="co">##  [13] "451742"    "737305"    "100615914" "456723"    "107967644" "454362"   </span></span>
<span><span class="co">##  [19] "464334"    "743667"    "741867"    "449586"    "100614256" "741708"   </span></span>
<span><span class="co">##  [25] "459164"    "746692"    "473976"    "452433"    "468889"    "745443"   </span></span>
<span><span class="co">##  [31] "460926"    "455644"    "451116"    "454684"    "744890"    "461229"   </span></span>
<span><span class="co">##  [37] "740513"    "104005232" "463949"    "469319"    "748673"    "450673"   </span></span>
<span><span class="co">##  [43] "468605"    "471455"    "456837"    "464611"    "452659"    "472079"   </span></span>
<span><span class="co">##  [49] "452307"    "454118"    "100616508" "465727"    "742828"    "737945"   </span></span>
<span><span class="co">##  [55] "107976794" "107976794" "746229"    "472893"    "456557"    "457056"   </span></span>
<span><span class="co">##  [61] "747265"    "736777"    "464460"    "451393"    "745691"    "454512"   </span></span>
<span><span class="co">##  [67] "466780"    "463861"    "744984"    "452566"    "457117"    NA         </span></span>
<span><span class="co">##  [73] "747936"    "459360"    "461436"    "464353"    "464074"    "466651"   </span></span>
<span><span class="co">##  [79] "451984"    "456243"    "464255"    "467738"    "466732"    "461244"   </span></span>
<span><span class="co">##  [85] "456929"    "460520"    "450562"    "450738"    "464140"    "459670"   </span></span>
<span><span class="co">##  [91] "452976"    "471703"    "741876"    "471135"    "461424"    "459828"   </span></span>
<span><span class="co">##  [97] "452295"    "460113"    "453565"    "741179"    "747276"    "470423"   </span></span>
<span><span class="co">## [103] "451967"    "450290"    "473975"    "473975"    "460157"    "462946"   </span></span>
<span><span class="co">## [109] "449638"    "738797"    "456076"    "451807"    "464031"    "739986"   </span></span>
<span><span class="co">## [115] "459173"    "460872"    "463484"    "462853"    "739167"    "457477"   </span></span>
<span><span class="co">## [121] "742027"    "746245"    "472764"    "747387"    "744096"    "101057233"</span></span>
<span><span class="co">## [127] "744811"    "463686"    "744435"    "468748"    "451175"    "460227"   </span></span>
<span><span class="co">## [133] "454744"    "739996"    NA          "450735"    "454478"    "457929"   </span></span>
<span><span class="co">## [139] "738397"    "458602"    "456908"    "451591"    "450310"    "107970333"</span></span>
<span><span class="co">## [145] "465012"    "463481"    "463481"    "460178"    "470365"    "742092"   </span></span>
<span><span class="co">## [151] "741184"    "459094"    "459374"    "456940"    "745779"    "454531"   </span></span>
<span><span class="co">## [157] "737918"    "107973114" "742100"    "470420"    "468499"    "467657"   </span></span>
<span><span class="co">## [163] "100608935" "462416"    "451281"    "470281"    "470281"    "452359"   </span></span>
<span><span class="co">## [169] "456862"    "456526"    "747462"    "474051"    "456155"    "458647"   </span></span>
<span><span class="co">## [175] "744390"    "455841"    "459096"    "459031"    "450574"    "449637"   </span></span>
<span><span class="co">## [181] "450628"    "470477"    "471247"    "453405"    "739128"    "454681"   </span></span>
<span><span class="co">## [187] "464707"    "470417"    "450933"    "459685"    "460443"    "468406"   </span></span>
<span><span class="co">## [193] "458803"    "467151"    "464550"    "745004"    "451416"    "735808"   </span></span>
<span><span class="co">## [199] "743144"    "460348"    "107974864" "471631"    "741897"    "463489"   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## $HALLMARK_ALLOGRAFT_REJECTION</span></span>
<span><span class="co">##   [1] "454210"    "461523"    "450363"    "100609296" "459646"    "740898"   </span></span>
<span><span class="co">##   [7] "466415"    "450170"    "465345"    "456984"    "744209"    "449497"   </span></span>
<span><span class="co">##  [13] "100608992" "459361"    "741390"    "468208"    "748142"    "473220"   </span></span>
<span><span class="co">##  [19] "748205"    "736543"    "454593"    "747004"    "454579"    "747123"   </span></span>
<span><span class="co">##  [25] "462689"    "460323"    "740071"    "450128"    "469524"    "449512"   </span></span>
<span><span class="co">##  [31] "748272"    "470617"    "451584"    "742330"    "451585"    "450124"   </span></span>
<span><span class="co">##  [37] "100615583" "473802"    "460569"    "745293"    "462191"    "740560"   </span></span>
<span><span class="co">##  [43] "470892"    "470900"    "735755"    "470426"    "460577"    "465021"   </span></span>
<span><span class="co">##  [49] "465607"    "736196"    "457127"    "453745"    "457277"    "738275"   </span></span>
<span><span class="co">##  [55] "471200"    NA          "459634"    "457770"    "469142"    "463415"   </span></span>
<span><span class="co">##  [61] "466216"    "458797"    "453714"    "469204"    "750603"    NA         </span></span>
<span><span class="co">##  [67] "100615835" "740028"    "451695"    "451158"    "471510"    "738331"   </span></span>
<span><span class="co">##  [73] "469584"    "739516"    "465940"    "461906"    "468521"    "457003"   </span></span>
<span><span class="co">##  [79] "472959"    "457020"    "467610"    "461873"    "452825"    "460623"   </span></span>
<span><span class="co">##  [85] "463280"    "746195"    "750725"    "100608816" "449592"    "471979"   </span></span>
<span><span class="co">##  [91] "471977"    "471974"    "462591"    "462540"    "494187"    "450196"   </span></span>
<span><span class="co">##  [97] "474132"    "473965"    "449517"    "747276"    "470077"    "743102"   </span></span>
<span><span class="co">## [103] "472749"    "469657"    "736204"    "460816"    "471723"    "455851"   </span></span>
<span><span class="co">## [109] "449564"    "737808"    "449644"    "739011"    "744277"    "450200"   </span></span>
<span><span class="co">## [115] "461472"    "456370"    "450884"    "470203"    "101059843" "449565"   </span></span>
<span><span class="co">## [121] "454005"    "463288"    "464245"    "745517"    "463371"    "470524"   </span></span>
<span><span class="co">## [127] "462386"    "450927"    "454294"    "454045"    "458607"    "735556"   </span></span>
<span><span class="co">## [133] "464979"    "450156"    "738375"    "456715"    "471734"    "736309"   </span></span>
<span><span class="co">## [139] "744486"    "459682"    "745667"    "472771"    "462888"    "748652"   </span></span>
<span><span class="co">## [145] "449582"    "458294"    "460699"    "459239"    "741196"    "460720"   </span></span>
<span><span class="co">## [151] "100322885" "469743"    "455026"    "740704"    "740477"    "450512"   </span></span>
<span><span class="co">## [157] "453993"    "456276"    "743176"    "748032"    "457607"    "462249"   </span></span>
<span><span class="co">## [163] "464277"    "737451"    "746600"    "737526"    "456065"    "461536"   </span></span>
<span><span class="co">## [169] "107966305" "746721"    "737070"    "459209"    "451169"    "450503"   </span></span>
<span><span class="co">## [175] "461971"    "461023"    "459834"    "100610925" "471978"    "746399"   </span></span>
<span><span class="co">## [181] "746814"    "456060"    "457742"    "451611"    "107971092" "461167"   </span></span>
<span><span class="co">## [187] "471325"    "471374"    "471167"    "494186"    "748737"    "464876"   </span></span>
<span><span class="co">## [193] "741922"    "745141"    "452125"    "453161"    "743187"    "459427"</span></span></code></pre>
<p>Now we can perform the local GREAT analysis.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">h_gene_sets</span>, <span class="st">"panTro6"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="use-ncbi-genomes-and-gene-sets">Use NCBI genomes and gene sets<a class="anchor" aria-label="anchor" href="#use-ncbi-genomes-and-gene-sets"></a>
</h2>
<p>NCBI provides genome data for a huge number of organisms. It is quite easy to obtain the GO gene sets for a specific organism with the <strong>AnnotationHub</strong> package where NCBI is also the data source. Users please go to the documentation of <strong>AnnotationHub</strong> to find out how to get GO gene sets with it.</p>
<p>It is relatively more difficult to obtain gene coordinates for a specific organism. Here the function <code><a href="../reference/getGenomeDataFromNCBI.html">getGenomeDataFromNCBI()</a></code> accepts a RefSeq accession number for a given genome assembly and returns information of the genome as well as the genes.</p>
<p>The RefSeq accession number can be obtained from <a href="https://www.ncbi.nlm.nih.gov/data-hub/genome/" class="external-link uri">https://www.ncbi.nlm.nih.gov/data-hub/genome/</a>. Here we take dolphin as an example. Its accession number is “GCF_011762595.1” and its genome page is at <a href="https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_011762595.1/" class="external-link uri">https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_011762595.1/</a>. We can get its genes with the command:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/getGenomeDataFromNCBI.html">getGenomeDataFromNCBI</a></span><span class="op">(</span><span class="st">"GCF_011762595.1"</span>, return_granges <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">genes</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 19081 ranges and 1 metadata column:</span></span>
<span><span class="co">##           seqnames        ranges strand |     gene_id</span></span>
<span><span class="co">##              &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##       [1]        1 234018-239154      + |   101335869</span></span>
<span><span class="co">##       [2]        1 241768-255021      + |   101336161</span></span>
<span><span class="co">##       [3]        1 259970-287776      + |   101336448</span></span>
<span><span class="co">##       [4]        1 288745-298452      + |   101336738</span></span>
<span><span class="co">##       [5]        1 325391-327082      - |   101321052</span></span>
<span><span class="co">##       ...      ...           ...    ... .         ...</span></span>
<span><span class="co">##   [19077]       MT    9921-10217      + |     7412053</span></span>
<span><span class="co">##   [19078]       MT   10211-11588      + |     7412054</span></span>
<span><span class="co">##   [19079]       MT   11790-13610      + |     7412055</span></span>
<span><span class="co">##   [19080]       MT   13594-14121      - |     7412056</span></span>
<span><span class="co">##   [19081]       MT   14195-15334      + |     7412057</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 23 sequences from an unspecified genome</span></span></code></pre>
<p>The chromosome length information is also included in <code>genes</code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">seqlengths</span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         1         2         3         4         5         6         7         8 </span></span>
<span><span class="co">## 183742880 176472748 171421769 144183200 137487549 115599616 114015307 108430135 </span></span>
<span><span class="co">##         9        10        11        12        13        14        15        16 </span></span>
<span><span class="co">## 105335178 102812509 102167733  89330967  88485069  88483154  86293990  83508544 </span></span>
<span><span class="co">##        17        18        19        20        21         X        MT </span></span>
<span><span class="co">##  78442772  78151963  58636127  58467291  35682248 128693445     16388</span></span></code></pre>
<p>Now you can send <code>genes</code> to <code><a href="../reference/great.html">great()</a></code> to perform the enrichment analysis.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">your_go_gene_sets</span>, tss_source <span class="op">=</span> <span class="va">genes</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
<p>The <code>GRanges</code> object can only be constructd if the genome is assembled on the “chromosome-level”. If it is not, e.g. on the scaffold or on the contig level. <code><a href="../reference/getGenomeDataFromNCBI.html">getGenomeDataFromNCBI()</a></code> returns the raw output which includes two data frames, one for the genome, and one for the genes. Let’s still take “GCF_011762595.1” as an example, but set <code>return_granges</code> to <code>FALSE</code> (which is the default value in the function):</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lt</span> <span class="op">=</span> <span class="fu"><a href="../reference/getGenomeDataFromNCBI.html">getGenomeDataFromNCBI</a></span><span class="op">(</span><span class="st">"GCF_011762595.1"</span>, return_granges <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">lt</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "genome" "gene"</span></span></code></pre>
<p>The first several rows in the two data frames.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lt</span><span class="op">$</span><span class="va">genome</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   genbankAccession refseqAccession chrName ucscStyleName    length</span></span>
<span><span class="co">## 1       CM022274.1     NC_047034.1       1          &lt;NA&gt; 183742880</span></span>
<span><span class="co">## 2       CM022275.1     NC_047035.1       2          &lt;NA&gt; 176472748</span></span>
<span><span class="co">## 3       CM022276.1     NC_047036.1       3          &lt;NA&gt; 171421769</span></span>
<span><span class="co">## 4       CM022277.1     NC_047037.1       4          &lt;NA&gt; 144183200</span></span>
<span><span class="co">## 5       CM022278.1     NC_047038.1       5          &lt;NA&gt; 137487549</span></span>
<span><span class="co">## 6       CM022279.1     NC_047039.1       6          &lt;NA&gt; 115599616</span></span></code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">lt</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   RefSeq Contig Accession  Start    End Strand EntreZ ID</span></span>
<span><span class="co">## 1             NC_047034.1 234018 239154      + 101335869</span></span>
<span><span class="co">## 2             NC_047034.1 241768 255021      + 101336161</span></span>
<span><span class="co">## 3             NC_047034.1 259970 287776      + 101336448</span></span>
<span><span class="co">## 4             NC_047034.1 288745 298452      + 101336738</span></span>
<span><span class="co">## 5             NC_047034.1 325391 327082      - 101321052</span></span>
<span><span class="co">## 6             NC_047034.1 337571 414483      + 101327235</span></span></code></pre>
<p>In the original data from NCBI, genes are associated to contigs, and that is why the first column in <code>lt$gene</code> contains RefSeq contig IDs. When the genome is assembled on the chromosome level, the contigs are chromosomes. For some organisms, there is no official chromosome name, then a specific contig ID from <code>lt$genome</code> can be used to map to the contig names.</p>
<p>In the following code, we assume in users’ input regions, the Genbank accession IDs are used for contigs, then we need to construct a gene <code>GRanges</code> object which also takes Genback accession IDs. This can be easily done as follows:</p>
<p>First we construct a mapping vector between Genbank IDs and RefSeq IDs:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">map</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="va">lt</span><span class="op">$</span><span class="va">genome</span><span class="op">$</span><span class="va">genbankAccession</span>, names <span class="op">=</span> <span class="va">lt</span><span class="op">$</span><span class="va">genome</span><span class="op">$</span><span class="va">refseqAccession</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">map</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  NC_047034.1  NC_047035.1  NC_047036.1  NC_047037.1  NC_047038.1  NC_047039.1 </span></span>
<span><span class="co">## "CM022274.1" "CM022275.1" "CM022276.1" "CM022277.1" "CM022278.1" "CM022279.1"</span></span></code></pre>
<p>Next apply <code>map</code> on the first column of <code>lt$gene</code> and then convert it into a <code>GRanges</code> object. Just make sure in the ID convertion, <code>NA</code> may be procudes and need to be removed.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">=</span> <span class="va">lt</span><span class="op">$</span><span class="va">gene</span></span>
<span><span class="va">genes</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">=</span> <span class="va">map</span><span class="op">[</span> <span class="va">genes</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">]</span></span>
<span><span class="va">genes</span> <span class="op">=</span> <span class="va">genes</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">genes</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="va">genes</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="va">genes</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, <span class="va">genes</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span>,</span>
<span>                strand <span class="op">=</span> <span class="va">genes</span><span class="op">[</span>, <span class="fl">4</span><span class="op">]</span>, gene_id <span class="op">=</span> <span class="va">genes</span><span class="op">[</span>, <span class="fl">5</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p>In <code>lt$genome</code> there is also contig length information. Simply construct it and set it to the <code>seqlengths</code> of <code>genes</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sl</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/structure.html" class="external-link">structure</a></span><span class="op">(</span><span class="va">lt</span><span class="op">$</span><span class="va">genome</span><span class="op">$</span><span class="va">length</span>, names <span class="op">=</span> <span class="va">lt</span><span class="op">$</span><span class="va">genome</span><span class="op">$</span><span class="va">genbankAccession</span><span class="op">)</span></span>
<span><span class="va">sl</span> <span class="op">=</span> <span class="va">sl</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sl</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">seqlengths</span><span class="op">(</span><span class="va">genes</span><span class="op">)</span> <span class="op">=</span> <span class="va">sl</span></span>
<span><span class="va">genes</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 19081 ranges and 1 metadata column:</span></span>
<span><span class="co">##             seqnames        ranges strand |     gene_id</span></span>
<span><span class="co">##                &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##       [1] CM022274.1 234018-239154      + |   101335869</span></span>
<span><span class="co">##       [2] CM022274.1 241768-255021      + |   101336161</span></span>
<span><span class="co">##       [3] CM022274.1 259970-287776      + |   101336448</span></span>
<span><span class="co">##       [4] CM022274.1 288745-298452      + |   101336738</span></span>
<span><span class="co">##       [5] CM022274.1 325391-327082      - |   101321052</span></span>
<span><span class="co">##       ...        ...           ...    ... .         ...</span></span>
<span><span class="co">##   [19077] EU557093.1    9921-10217      + |     7412053</span></span>
<span><span class="co">##   [19078] EU557093.1   10211-11588      + |     7412054</span></span>
<span><span class="co">##   [19079] EU557093.1   11790-13610      + |     7412055</span></span>
<span><span class="co">##   [19080] EU557093.1   13594-14121      - |     7412056</span></span>
<span><span class="co">##   [19081] EU557093.1   14195-15334      + |     7412057</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 23 sequences from an unspecified genome</span></span></code></pre>
<p>And similarly, <code>genes</code> can be directly sent to enrichment analysis.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">your_go_gene_sets</span>, tss_source <span class="op">=</span> <span class="va">genes</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
<p>Let’s demonstrate how to obtain GO gene sets for a less-studied organism from the <code>OrgDb</code> object. Still taking dolphin as an example, we can get its gene coordinates with the <code><a href="../reference/getGenomeDataFromNCBI.html">getGenomeDataFromNCBI()</a></code> function by specifying its GenBank accession number. The <code>OrgDb</code> object for dolphin can be obtained with the <strong>AnnotationHub</strong> package. First construct an <code>AnnotationHub</code> object:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">AnnotationHub</span><span class="op">)</span></span>
<span><span class="va">ah</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">AnnotationHub</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Next we need to search dolphin on Annotation hub. It is sometimes not easy to find the dataset from AnnotationHub as you want. In this example, we use query words of “truncatus” and “OrgDb” (unfortunately, the word “dolphin” can not find the correct file).</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">ah</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"truncatus"</span>, <span class="st">"OrgDb"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The returned message shows the ID of the OrgDb database for dolphin is “AH112418”. Then we can directly obtain the <code>OrgDb</code> object by:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">orgdb</span> <span class="op">=</span> <span class="va">ah</span><span class="op">[[</span><span class="st">"AH112418"</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p>With another helper function <code><a href="../reference/getGeneSetsFromOrgDb.html">getGeneSetsFromOrgDb()</a></code>, we can obtain the GO gene sets for dolphin:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">go_gs</span> <span class="op">=</span> <span class="fu"><a href="../reference/getGeneSetsFromOrgDb.html">getGeneSetsFromOrgDb</a></span><span class="op">(</span><span class="va">orgdb</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">go_gs</span>, tss_source <span class="op">=</span> <span class="va">genes</span>, <span class="va">...</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="use-kegg-gene-sets">Use KEGG gene sets<a class="anchor" aria-label="anchor" href="#use-kegg-gene-sets"></a>
</h2>
<p>KEGG supports a huge number of organisms. How to obtain the pathway gene sets for an organism has been introduced in the vignette <a href="other-geneset-databases.html#kegg-pathways">“Work with other geneset databases”</a>. Now the task is to obtain the gene/TSS definitions of that organism. On the KEGG database, each organism has a RefSeq accession ID associated and this ID can be found on the webpage of the organism. For example, the web page of turkey is <a href="https://www.genome.jp/kegg-bin/show_organism?org=mgp" class="external-link uri">https://www.genome.jp/kegg-bin/show_organism?org=mgp</a> where the RefSeq assembly access ID is “GCF_000146605.3”. This ID can be used later with <code><a href="../reference/getGenomeDataFromNCBI.html">getGenomeDataFromNCBI()</a></code> to get the gene coordinates of that genome assembly.</p>
<p><strong>rGREAT</strong> has a function <code><a href="../reference/getKEGGGenome.html">getKEGGGenome()</a></code> which automatically extracts the RefSeq accession ID for an organism:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getKEGGGenome.html">getKEGGGenome</a></span><span class="op">(</span><span class="st">"mgp"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GCF_000146605.3"</span></span></code></pre>
<p>Then, the following code performs GREAT analysis for any organism supported on KEGG, e.g. turkey:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="../reference/getGenomeDataFromNCBI.html">getGenomeDataFromNCBI</a></span><span class="op">(</span><span class="fu"><a href="../reference/getKEGGGenome.html">getKEGGGenome</a></span><span class="op">(</span><span class="st">"mgp"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">gene_sets</span> <span class="op">=</span> <span class="fu"><a href="../reference/getKEGGPathways.html">getKEGGPathways</a></span><span class="op">(</span><span class="st">"mgp"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">...</span>, <span class="va">gene_sets</span>, <span class="va">genes</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="self-define-tss-and-gene-sets">Self-define TSS and gene sets<a class="anchor" aria-label="anchor" href="#self-define-tss-and-gene-sets"></a>
</h2>
<p>Since <code><a href="../reference/great.html">great()</a></code> allows both self-defined TSS and gene sets, this means <code><a href="../reference/great.html">great()</a></code> can be independent to organisms. Please refer to the vignette <a href="local-GREAT.html#manually-set-gene-sets-and-transcriptome-annotations">“Analyze with local GREAT”</a> to find out how to manuallly set both TSS and gene sets.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
