<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyze with local GREAT • rGREAT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyze with local GREAT">
<meta property="og:description" content="rGREAT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rGREAT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.99.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/local-GREAT.html">Analyze with local GREAT</a>
    </li>
    <li>
      <a href="../articles/online-GREAT.html">Analyze with online GREAT</a>
    </li>
    <li>
      <a href="../articles/other-docs.html">Other documents</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/rGREAT/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyze with local GREAT</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu ( <a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a> )</h4>
            
            <h4 data-toc-skip class="date">2022-06-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/rGREAT/blob/HEAD/vignettes/local-GREAT.Rmd" class="external-link"><code>vignettes/local-GREAT.Rmd</code></a></small>
      <div class="hidden name"><code>local-GREAT.Rmd</code></div>

    </div>

    
    
<p><strong>GREAT</strong> (<a href="http://great.stanford.edu" class="external-link">Genomic Regions Enrichment of Annotations Tool</a>) is a popular web-based tool to associate biological functions to genomic regions, however, its nature of being an online tool has several limitations:</p>
<ol style="list-style-type: decimal">
<li>limited number of supported organisms. Current version 4.0.4 only supports four organisms: “hg38”, “hg19”, “mm10” and “mm9”.</li>
<li>limited number of ontologies (or gene set collections). Current version 4.0.4 only supports seven ontologies (<a href="https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655442/Version+History" class="external-link uri">https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655442/Version+History</a>).</li>
<li>maximal 500,000 test regions (<a href="https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655402/File+Size" class="external-link uri">https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655402/File+Size</a>).</li>
<li>the update of annotation databases is only controlled by the GREAT developers.</li>
</ol>
<p>From <strong>rGREAT</strong> version 1.99.0, it implements the GREAT algorithms locally and it can be seamlessly integrated to the Bioconductor annotation ecosystem. This means, theoretically with <strong>rGREAT</strong>, it is possible to perform GREAT analysis with any organism and with any type of gene set collection / ontology. Another advantage is, Bioconductor annotation packages are always very well maintained and updated, which means the data source of your analysis can be ensured to be most up-to-date.</p>
<p>First let’s load the <strong>rGREAT</strong> package and generate a random set of regions:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/rGREAT" class="external-link">rGREAT</a></span><span class="op">)</span></code></pre></div>
<div class="section level2">
<h2 id="perform-local-great-with-great">Perform local GREAT with <code>great()</code><a class="anchor" aria-label="anchor" href="#perform-local-great-with-great"></a>
</h2>
<p>The function <code><a href="../reference/great.html">great()</a></code> is the core function to perform local GREAT analysis. You can either use built-in annotations or use self-provided annotations.</p>
<div class="section level3">
<h3 id="with-defaultly-supported-annotation-packages">With defaultly supported annotation packages<a class="anchor" aria-label="anchor" href="#with-defaultly-supported-annotation-packages"></a>
</h3>
<p><code><a href="../reference/great.html">great()</a></code> has integrated many annotation databases which cover many organisms. Users simply specify the name of gene set collection (via the second argument) and the source of TSS (via the third argument).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">res</span> <span class="op">=</span> <span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"MSigDB:H"</span>, <span class="st">"txdb:hg19"</span><span class="op">)</span>
<span class="va">res</span></code></pre></div>
<pre><code><span class="co">## 1000 regions are associated to 1414 genes' extended TSSs.</span>
<span class="co">##   TSS source: TxDb.Hsapiens.UCSC.hg19.knownGene</span>
<span class="co">##   Genome: hg19</span>
<span class="co">##   OrgDb: org.Hs.eg.db</span>
<span class="co">##   Gene sets: MSigDB:H</span>
<span class="co">##   Background: whole genome</span>
<span class="co">## Mode: Basal plus extension</span>
<span class="co">##   Proximal: 5000 bp upstream, 1000 bp downstream,</span>
<span class="co">##   plus Distal: up to 1000000 bp</span></code></pre>
<p>There are following supported gene set collections. The first category are GO gene sets. The gene sets data is from <strong>GO.db</strong> package.</p>
<ul>
<li>
<code>"GO:BP"</code>: Biological Process.</li>
<li>
<code>"GO:CC"</code>: Cellular Component.</li>
<li>
<code>"GO:MP"</code>: Molecular Function.</li>
</ul>
<p>The prefix <code>GO:</code> can be omitted when it is specified in <code><a href="../reference/great.html">great()</a></code>.</p>
<p>The second category of gene sets are from <a href="https://www.gsea-msigdb.org/gsea/msigdb/" class="external-link">MSigDB</a>. Note this is only for human:</p>
<ul>
<li>
<code>"msigdb:H"</code> Hallmark gene sets.</li>
<li>
<code>"msigdb:C1"</code> Positional gene sets.</li>
<li>
<code>"msigdb:C2"</code> Curated gene sets.</li>
<li>
<code>"msigdb:C2:CGP"</code> C2 subcategory: chemical and genetic perturbations gene sets.</li>
<li>
<code>"msigdb:C2:CP"</code> C2 subcategory: canonical pathways gene sets.</li>
<li>
<code>"msigdb:C2:CP:BIOCARTA"</code> C2 subcategory: BioCarta subset of CP.</li>
<li>
<code>"msigdb:C2:CP:KEGG"</code> C2 subcategory: KEGG subset of CP.</li>
<li>
<code>"msigdb:C2:CP:PID"</code> C2 subcategory: PID subset of CP.</li>
<li>
<code>"msigdb:C2:CP:REACTOME"</code> C2 subcategory: REACTOME subset of CP.</li>
<li>
<code>"msigdb:C2:CP:WIKIPATHWAYS"</code> C2 subcategory: WIKIPATHWAYS subset of CP.</li>
<li>
<code>"msigdb:C3"</code> Regulatory target gene sets.</li>
<li>
<code>"msigdb:C3:MIR:MIRDB"</code> miRDB of microRNA targets gene sets.</li>
<li>
<code>"msigdb:C3:MIR:MIR_LEGACY"</code> MIR_Legacy of MIRDB.</li>
<li>
<code>"msigdb:C3:TFT:GTRD"</code> GTRD transcription factor targets gene sets.</li>
<li>
<code>"msigdb:C3:TFT:TFT_LEGACY"</code> TFT_Legacy.</li>
<li>
<code>"msigdb:C4"</code> Computational gene sets.</li>
<li>
<code>"msigdb:C4:CGN"</code> C4 subcategory: cancer gene neighborhoods gene sets.</li>
<li>
<code>"msigdb:C4:CM"</code> C4 subcategory: cancer modules gene sets.</li>
<li>
<code>"msigdb:C5"</code> Ontology gene sets.</li>
<li>
<code>"msigdb:C5:GO:BP"</code> C5 subcategory: BP subset.</li>
<li>
<code>"msigdb:C5:GO:CC"</code> C5 subcategory: CC subset.</li>
<li>
<code>"msigdb:C5:GO:MF"</code> C5 subcategory: MF subset.</li>
<li>
<code>"msigdb:C5:HPO"</code> C5 subcategory: human phenotype ontology gene sets.</li>
<li>
<code>"msigdb:C6"</code> Oncogenic signature gene sets.</li>
<li>
<code>"msigdb:C7"</code> Immunologic signature gene sets.</li>
<li>
<code>"msigdb:C7:IMMUNESIGDB"</code> ImmuneSigDB subset of C7.</li>
<li>
<code>"msigdb:C7:VAX"</code> C7 subcategory: vaccine response gene sets.</li>
<li>
<code>"msigdb:C8"</code> Cell type signature gene sets.</li>
</ul>
<p>The prefix <code>msigdb:</code> can be omitted when specified in <code><a href="../reference/great.html">great()</a></code> and the name of a MSigDb can be used as case insensitive.</p>
<p><strong>rGREAT</strong> supports TSS from several sources. The value of argument <code>tss_source</code> should be encoded in a special format:</p>
<ul>
<li>Name of a <code>TxDb.*</code> package, e.g. <code>TxDb.Hsapiens.UCSC.hg19.knownGene</code>. Supported packages are in <code>rGREAT:::BIOC_ANNO_PKGS$txdb</code>.</li>
<li>Genome version of the organism, e.g. “hg19”. Then the corresponding TxDb package will be used.</li>
<li>In a format of <code>RefSeq:$genome</code> where <code>$genome</code> is the genome version of an organism. RefSeqSelect genes will be used.</li>
<li>In a format of <code>RefSeqCurated:$genome</code> where <code>$genome</code> is the genome version of an organism. RefSeqCurated subset will be used.</li>
<li>In a format of <code>RefSeqSelect:$genome</code> where <code>$genome</code> is the genome version of an organism. RefSeqSelect subset will be used.</li>
<li>In a format of <code>Gencode_v$version</code> where <code>$version</code> is Gencode version, such as 19 (for human) or M21 for mouse. Gencode protein coding genes will be used.</li>
<li>In a format of <code>GREAT:$genome</code>, where <code>$genome</code> can only be “mm9”, “mm10”, “hg19”, “hg38”. TSS from GREAT will be used. The data is downloaded from <a href="https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655445/Genes" class="external-link uri">https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655445/Genes</a>.</li>
</ul>
<p>The difference of <code>RefSeqCurated</code> and <code>RefSeqSelect</code> is explained in <a href="https://genome.ucsc.edu/cgi-bin/hgTrackUi?db=hg38&amp;g=refSeqComposite" class="external-link uri">https://genome.ucsc.edu/cgi-bin/hgTrackUi?db=hg38&amp;g=refSeqComposite</a>.</p>
<p>Some examples are:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"hg19"</span><span class="op">)</span>
<span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"TxDb.Hsapiens.UCSC.hg19.knownGene"</span><span class="op">)</span>
<span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"RefSeq:hg19"</span><span class="op">)</span>
<span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"GREAT:hg19"</span><span class="op">)</span>
<span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"Gencode_v19"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="manually-set-gene-sets">Manually set gene sets<a class="anchor" aria-label="anchor" href="#manually-set-gene-sets"></a>
</h3>
<p>If users have their own gene sets, the gene sets can be set as a named list of vectors where each vector corresponds to one gene set. Please note the genes in the gene sets <strong>must be in Entrez ID type</strong>. In the following example, we use a gene set collection from <a href="http://dsigdb.tanlab.org/DSigDBv1.0/collection.html" class="external-link">DSigDB</a>. This collection (FDA approved kinase inhibitors) only contains 28 gene sets.</p>
<p>Here <code><a href="../reference/read_gmt.html">read_gmt()</a></code> is a simple function that reads a gmt file as a list of vectors, also performs gene ID conversion.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gs</span> <span class="op">=</span> <span class="fu"><a href="../reference/read_gmt.html">read_gmt</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">url</a></span><span class="op">(</span><span class="st">"http://dsigdb.tanlab.org/Downloads/D2_LINCS.gmt"</span><span class="op">)</span>, 
    from <span class="op">=</span> <span class="st">"SYMBOL"</span>, to <span class="op">=</span> <span class="st">"ENTREZ"</span>, orgdb <span class="op">=</span> <span class="st">"org.Hs.eg.db"</span><span class="op">)</span>
<span class="va">gs</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span></code></pre></div>
<pre><code><span class="co">## $GSK429286A</span>
<span class="co">##  [1] "5566"  "9475"  "81788" "6093"  "5562"  "26524" "5592"  "5979"  "640"  </span>
<span class="co">## [10] "5567"  "3717"  "5613"  "23012" "695"   "3718" </span>
<span class="co">## </span>
<span class="co">## $`BS-181`</span>
<span class="co">##  [1] "122011" "1452"   "1022"   "7272"   "1454"   "1453"   "65975"  "147746"</span>
<span class="co">##  [9] "1859"   "1195"   "1196"   "9149"   "57396"  "5261"</span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="va">gs</span>, <span class="st">"hg19"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 1000 regions are associated to 1414 genes' extended TSSs.</span>
<span class="co">##   TSS source: TxDb.Hsapiens.UCSC.hg19.knownGene</span>
<span class="co">##   Genome: hg19</span>
<span class="co">##   OrgDb: org.Hs.eg.db</span>
<span class="co">##   Gene sets: self-provided</span>
<span class="co">##   Background: whole genome</span>
<span class="co">## Mode: Basal plus extension</span>
<span class="co">##   Proximal: 5000 bp upstream, 1000 bp downstream,</span>
<span class="co">##   plus Distal: up to 1000000 bp</span></code></pre>
</div>
<div class="section level3">
<h3 id="manually-set-tss">Manually set TSS<a class="anchor" aria-label="anchor" href="#manually-set-tss"></a>
</h3>
<p>Users may have their own set of genes/TSS, as in the following example:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/connections.html" class="external-link">gzfile</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"GREATv4.genes.hg19.tsv.gz"</span>, package <span class="op">=</span> <span class="st">"rGREAT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co"># note there must be a 'gene_id' column</span>
<span class="va">tss</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="va">df</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span>, <span class="va">df</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span>, 
    strand <span class="op">=</span> <span class="va">df</span><span class="op">[</span>, <span class="fl">4</span><span class="op">]</span>, gene_id <span class="op">=</span> <span class="va">df</span><span class="op">[</span>, <span class="fl">5</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tss</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 6 ranges and 1 metadata column:</span>
<span class="co">##       seqnames    ranges strand |     gene_id</span>
<span class="co">##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">##   [1]     chr1     69090      + |       OR4F5</span>
<span class="co">##   [2]     chr1    367639      + |      OR4F29</span>
<span class="co">##   [3]     chr1    622053      - |      OR4F16</span>
<span class="co">##   [4]     chr1    861117      + |      SAMD11</span>
<span class="co">##   [5]     chr1    894670      - |       NOC2L</span>
<span class="co">##   [6]     chr1    895966      + |      KLHL17</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 25 sequences from an unspecified genome; no seqlengths</span></code></pre>
<p>In this case, users must manually generate an “extended TSS” by <code><a href="../reference/extendTSS.html">extendTSS()</a></code> function. They should also explicitly specify the gene ID type in <code><a href="../reference/extendTSS.html">extendTSS()</a></code> so that <code><a href="../reference/great.html">great()</a></code> can correctly map to the genes in gene sets.</p>
<p>In the example, IDs for genes in <code>tss</code> are symbols, thus, <code>gene_id_type</code> must be set to <code>"SYMBOL"</code> so that the correct gene ID type will be selected for internal gene sets.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">et</span> <span class="op">=</span> <span class="fu"><a href="../reference/extendTSS.html">extendTSS</a></span><span class="op">(</span><span class="va">tss</span>, genome <span class="op">=</span> <span class="st">"hg19"</span>, gene_id_type <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span>
<span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"msigdb:h"</span>, extended_tss <span class="op">=</span> <span class="va">et</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 1000 regions are associated to 1419 genes' extended TSSs.</span>
<span class="co">##   TSS source: self-provided</span>
<span class="co">##   Genome: hg19</span>
<span class="co">##   OrgDb: org.Hs.eg.db</span>
<span class="co">##   Gene sets: msigdb:h</span>
<span class="co">##   Background: whole genome</span>
<span class="co">## Mode: Basal plus extension</span>
<span class="co">##   Proximal: 5000 bp upstream, 1000 bp downstream,</span>
<span class="co">##   plus Distal: up to 1000000 bp</span></code></pre>
<p>If gene ID type in <code>tss</code> is one of Ensembl/RefSeq/Entrez ID, <code>gene_id_type</code> argument can be omitted because the ID type can be automatically inferred from the format of the gene IDs, but it is always a good idea to explicitly specify it if the data is self-provided.</p>
</div>
<div class="section level3">
<h3 id="manually-set-gene-sets-and-transcriptome-annotations">Manually set gene sets and transcriptome annotations<a class="anchor" aria-label="anchor" href="#manually-set-gene-sets-and-transcriptome-annotations"></a>
</h3>
<p>If your organism is not defaultly supported, you can always use the <code><a href="../reference/extendTSS.html">extendTSS()</a></code> to manually construct one. Note in GREAT algorithm, TSS are first extended by a rule (e.g. basal plus extension). <code><a href="../reference/extendTSS.html">extendTSS()</a></code> accepts a <code>GRanges</code> object of gene or TSS and it returns a new <code>GRanges</code> object of extended TSS.</p>
<p>In the following example, since I don’t have data for the organism not defaultly supported by <strong>rGREAT</strong>. I simply use human data to demonstrate how to manually construct the extended TSS.</p>
<p>There are two objects for <code><a href="../reference/extendTSS.html">extendTSS()</a></code>: the gene (or the TSS) and the length of chromosomes. The gene object must have a meta column named “gene_id” which stores gene ID in a specific type (this ID type will be mapped to the genes in gene sets). The chromosome length object is a named vector and it also controls the total set of chromosomes to be used in the analysis.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="va">gene</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicFeatures/man/transcripts.html" class="external-link">genes</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg19.knownGene</span><span class="op">)</span>
<span class="va">gene</span> <span class="op">=</span> <span class="va">gene</span><span class="op">[</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 6 ranges and 1 metadata column:</span>
<span class="co">##             seqnames              ranges strand |     gene_id</span>
<span class="co">##                &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span>
<span class="co">##           1    chr19   58858172-58874214      - |           1</span>
<span class="co">##          10     chr8   18248755-18258723      + |          10</span>
<span class="co">##         100    chr20   43248163-43280376      - |         100</span>
<span class="co">##        1000    chr18   25530930-25757445      - |        1000</span>
<span class="co">##       10000     chr1 243651535-244006886      - |       10000</span>
<span class="co">##   100008586     chrX   49217763-49233491      + |   100008586</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 93 sequences (1 circular) from hg19 genome</span></code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gl</span> <span class="op">=</span> <span class="fu">seqlengths</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>  <span class="co"># restrict to normal chromosomes</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gl</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##      chr1      chr2      chr3      chr4      chr5      chr6 </span>
<span class="co">## 249250621 243199373 198022430 191154276 180915260 171115067</span></code></pre>
<p>Simply send <code>gene</code> and <code>gl</code> to <code><a href="../reference/extendTSS.html">extendTSS()</a></code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">et</span> <span class="op">=</span> <span class="fu"><a href="../reference/extendTSS.html">extendTSS</a></span><span class="op">(</span><span class="va">gene</span>, <span class="va">gl</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">et</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 6 ranges and 3 metadata columns:</span>
<span class="co">##             seqnames        ranges strand |     gene_id tss_position</span>
<span class="co">##                &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;integer&gt;</span>
<span class="co">##   100287102     chr1       1-28962      * |   100287102        11874</span>
<span class="co">##      653635     chr1   12873-64091      * |      653635        29961</span>
<span class="co">##       79501     chr1  34961-139567      * |       79501        69091</span>
<span class="co">##      729737     chr1  70090-713069      * |      729737       140566</span>
<span class="co">##   100288069     chr1 145566-757971      * |   100288069       714068</span>
<span class="co">##      643837     chr1 719068-763970      * |      643837       762971</span>
<span class="co">##              tss_strand</span>
<span class="co">##             &lt;character&gt;</span>
<span class="co">##   100287102           +</span>
<span class="co">##      653635           -</span>
<span class="co">##       79501           +</span>
<span class="co">##      729737           -</span>
<span class="co">##   100288069           -</span>
<span class="co">##      643837           +</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 24 sequences from an unspecified genome</span></code></pre>
<p>We can also manually construct the gene sets object, which is simply a named list of vectors where each vector contains genes in a gene set. Here we directly use the object <code>gs</code> generated before which is a gene set collection from DSigDB.</p>
<p><strong>Please note again, gene IDs in <code>gs</code> should be the same as in <code>et</code>.</strong></p>
<p>Now <code>gs</code> and <code>et</code> can be sent to <code><a href="../reference/great.html">great()</a></code> to perform local GREAT with the annotation data you manually provided.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, gene_sets <span class="op">=</span> <span class="va">gs</span>, extended_tss <span class="op">=</span> <span class="va">et</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## 1000 regions are associated to 1490 genes' extended TSSs.</span>
<span class="co">##   TSS source: self-provided</span>
<span class="co">##   Genome: unknown</span>
<span class="co">##   Gene sets: self-provided</span>
<span class="co">##   Background: whole genome</span>
<span class="co">## Mode: Basal plus extension</span>
<span class="co">##   Proximal: 5000 bp upstream, 1000 bp downstream,</span>
<span class="co">##   plus Distal: up to 1000000 bp</span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="use-biomart-genes-and-genesets">Use BioMart genes and genesets<a class="anchor" aria-label="anchor" href="#use-biomart-genes-and-genesets"></a>
</h2>
<p><strong>rGREAT</strong> also supports GO gene sets for a huge number of organisms retrieved from <a href="https://www.ensembl.org/info/data/biomart/index.html" class="external-link">Ensembl BioMart</a>. A specific organism can be set with the <code>biomart_dataset</code> argument:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># giant panda</span>
<span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, biomart_dataset <span class="op">=</span> <span class="st">"amelanoleuca_gene_ensembl"</span><span class="op">)</span></code></pre></div>
<p>Note both TSS and gene sets are from BioMart. The value for gene sets (the second argument) can only be one of <code>"GO:BP"</code>, <code>"GO:CC"</code> and <code>"GO:MF"</code>.</p>
<p><strong>rGREAT</strong> now supports 562 organisms. The complete list can be found at <a href="https://jokergoo.github.io/rGREAT_genesets/" class="external-link uri">https://jokergoo.github.io/rGREAT_genesets/</a>. Just make sure the genome version is the same as the genome of your input regions.</p>
</div>
<div class="section level2">
<h2 id="set-background-regions">Set background regions<a class="anchor" aria-label="anchor" href="#set-background-regions"></a>
</h2>
<p>In the online GREAT tool, if background regions are set, <a href="https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655456/Statistics#Statistics-ExplicitBackgroundSet" class="external-link">it actually uses a different test for the enrichment analysis</a>. In GREAT, when background is set, input regions should be <strong>exactly subset of background</strong>. For example, let’s say a background region list contains five regions: <code>[1, 10], [15, 23], [34, 38], [40, 49], [54, 63]</code>, input regions can only be a subset of the five regions, which means they can take <code>[15, 23], [40, 49]</code>, but it cannot take <code>[16, 20], [39, 51]</code>. In this setting, regions are taken as single units and Fisher’s exact test is applied for calculating the enrichment (by testing number of regions in the 2x2 contigency table).</p>
<p>This might be useful for certain cases, e.g., for a specific transcriptional factor (TF), we take the union of ChIP-seq peaks of this TF from all tissues as the background set, and only take peaks from one specific tisse as input region set, and we want to test the enrichment of TF peaks in the tissue compared to the “background”. However, this “background definition” might not be the case as many other users may think. They might take “background” as a set of regions where they only want to perform GREAT (the Binomial method) in. E.g. they may want to exclude “gap regions / unsequenced regions” from the analysis because the null hypothesis of Binomial test is the input regions are uniformly distributed in the genome. Since the unsequenced regions will never be measured, they should be excluded from the analysis. Other examples are that the background can be set as regions showing similar GC contents or CpG density as the input regions.</p>
<p><code><a href="../reference/great.html">great()</a></code> supports two arguments <code>background</code> and <code>exclude</code> for setting a proper background. If any one of the two is set, the input regions and the extended TSS regions are intersected to the background, and GREAT algorithm is only applied to the reduced regions.</p>
<p>When whole genome is set as background, denote <span class="math inline">\(N_1\)</span> as the total number of input regions, <span class="math inline">\(p_1\)</span> as the fraction of genome that are overlapped to extended TSS of genes in a certain gene set, and <span class="math inline">\(K_1\)</span> as the number of regions that overlap to the gene set associated regions, then the enrichment test is based on the random variable <span class="math inline">\(K_1\)</span> which follows Binomial distribution <span class="math inline">\(K_1 \sim B(p_1, N_1)\)</span>.</p>
<p>Similarly, when background regions are set, denote <span class="math inline">\(N_2\)</span> as the total number of input regions that <u>overlap to backgroud</u>, <span class="math inline">\(p_2\)</span> as <u>the fraction of background</u> that are overlapped to extended TSS of genes in a certain gene set, and <span class="math inline">\(K_2\)</span> as the number of regions that overlap to the gene set associated regions <u>and also overlap to background</u>, then the enrichment test is based on the random variable <span class="math inline">\(K_2\)</span> which follows Binomial distribution <span class="math inline">\(K_2 \sim B(p_2, N_2)\)</span>.</p>
<p>In fact, the native hypergeometric method in GREAT can be approximated to the binomial method here. Nevertheless, the binomial method is more general and it has no restriction as the hypergeometric method where input regions must be perfect subsets of backgrounds.</p>
<p>In the following example, <code><a href="../reference/getGapFromUCSC.html">getGapFromUCSC()</a></code> can be used to retrieve gap regions from UCSC table browser.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gap</span> <span class="op">=</span> <span class="fu"><a href="../reference/getGapFromUCSC.html">getGapFromUCSC</a></span><span class="op">(</span><span class="st">"hg19"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"MSigDB:H"</span>, <span class="st">"hg19"</span>, exclude <span class="op">=</span> <span class="va">gap</span><span class="op">)</span></code></pre></div>
<p>Alternatively, <code>background</code> and <code>exclude</code> can also be set to a vector of chromosome names, then the whole selected chromosomes will be included/excluded from the analysis.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, background <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, exclude <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"chrX"</span>, <span class="st">"chrY"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="get-enrichment-table">Get enrichment table<a class="anchor" aria-label="anchor" href="#get-enrichment-table"></a>
</h2>
<p>Simply use <code><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable()</a></code> function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tb</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tb</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">##                                   id genome_fraction observed_region_hits</span>
<span class="co">## 1           HALLMARK_SPERMATOGENESIS     0.013006202                   21</span>
<span class="co">## 2                 HALLMARK_APOPTOSIS     0.014674462                   23</span>
<span class="co">## 3       HALLMARK_IL2_STAT5_SIGNALING     0.019680928                   25</span>
<span class="co">## 4      HALLMARK_BILE_ACID_METABOLISM     0.009296570                   13</span>
<span class="co">## 5        HALLMARK_HEDGEHOG_SIGNALING     0.006454074                    9</span>
<span class="co">## 6 HALLMARK_OXIDATIVE_PHOSPHORYLATION     0.011548106                   14</span>
<span class="co">##   fold_enrichment    p_value  p_adjust mean_tss_dist observed_gene_hits</span>
<span class="co">## 1        1.614614 0.02431695 0.5262531        261761                 19</span>
<span class="co">## 2        1.567349 0.02567088 0.5262531        248493                 18</span>
<span class="co">## 3        1.270265 0.13732116 0.8653711        189866                 22</span>
<span class="co">## 4        1.398365 0.14585046 0.8653711        167309                 12</span>
<span class="co">## 5        1.394468 0.20240993 0.8653711        307394                  7</span>
<span class="co">## 6        1.212320 0.27093709 0.8653711        148492                 13</span>
<span class="co">##   gene_set_size fold_enrichment_hyper p_value_hyper p_adjust_hyper</span>
<span class="co">## 1           134             1.8620353   0.006231823      0.1277524</span>
<span class="co">## 2           158             1.4960790   0.055874652      0.3818101</span>
<span class="co">## 3           196             1.4740279   0.043234738      0.3716015</span>
<span class="co">## 4           112             1.4070267   0.144591553      0.5928254</span>
<span class="co">## 5            36             2.5534928   0.017302502      0.2364675</span>
<span class="co">## 6           199             0.8578856   0.756322786      0.9803129</span></code></pre>
<p>In <code><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable()</a></code>, you can set argument <code>min_region_hits</code> to set the minimal number of input regions that hit a geneset-associated regions. Note the adjusted p-values will be recalculated in the table.</p>
<p>There are also columns for hypergeometric test on the numbers of genes, the same as in the original GREAT method.</p>
<p style="color:red">
There is a new column “mean_tss_dist” in the result table which is the mean absolute distance of input regions to TSS of genes in a gene set. Please note with larger distance to TSS, the more we need to be careful with the reliability of the associations between input regions to genes.
</p>
</div>
<div class="section level2">
<h2 id="make-volcano-plot">Make volcano plot<a class="anchor" aria-label="anchor" href="#make-volcano-plot"></a>
</h2>
<p>In differential gene expression analysis, volcano plot is used to visualize relations between log2 fold change and (adjusted) p-values. Similarly, we can also use volcano plot to visualize relations between fold enrichment and (adjusted) p-values for the enrichment analysis. The plot is made by the function <code><a href="../reference/plotVolcano-dispatch.html">plotVolcano()</a></code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotVolcano-dispatch.html">plotVolcano</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></code></pre></div>
<p><img src="local-GREAT_files/figure-html/unnamed-chunk-15-1.png" width="576" style="display: block; margin: auto;"></p>
<p>As the enrichment analysis basically only looks for over-representations, it is actually half volcano.</p>
</div>
<div class="section level2">
<h2 id="get-region-gene-associations">Get region-gene associations<a class="anchor" aria-label="anchor" href="#get-region-gene-associations"></a>
</h2>
<p><code><a href="../reference/plotRegionGeneAssociations-dispatch.html">plotRegionGeneAssociations()</a></code> generates three plots similar as those by online GREAT. <code><a href="../reference/getRegionGeneAssociations-dispatch.html">getRegionGeneAssociations()</a></code> returns a <code>GRanges</code> object containing associations between regions and genes.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotRegionGeneAssociations-dispatch.html">plotRegionGeneAssociations</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></code></pre></div>
<p><img src="local-GREAT_files/figure-html/unnamed-chunk-16-1.png" width="960" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getRegionGeneAssociations-dispatch.html">getRegionGeneAssociations</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 884 ranges and 2 metadata columns:</span>
<span class="co">##         seqnames            ranges strand | annotated_genes    dist_to_TSS</span>
<span class="co">##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;  &lt;IntegerList&gt;</span>
<span class="co">##     [1]     chr1   9204434-9208784      * |     GPR157,H6PD  -15205,-86079</span>
<span class="co">##     [2]     chr1   9853594-9859363      * |   PIK3CD,CLSTN1   141804,25187</span>
<span class="co">##     [3]     chr1 10862809-10871681      * |  CASZ1,C1orf127   -6076,170413</span>
<span class="co">##     [4]     chr1 12716970-12723206      * | AADACL4,AADACL3   12404,-52912</span>
<span class="co">##     [5]     chr1 13814692-13823250      * | PRAMEF17,LRRC38    98604,16992</span>
<span class="co">##     ...      ...               ...    ... .             ...            ...</span>
<span class="co">##   [880]     chrY 16384799-16389852      * |   TMSB4Y,NLGN4Y 569352,-244636</span>
<span class="co">##   [881]     chrY 16842593-16846893      * |          NLGN4Y         208105</span>
<span class="co">##   [882]     chrY 16998750-17003225      * |          NLGN4Y         364262</span>
<span class="co">##   [883]     chrY 24401612-24406252      * |   RBMY1E,RBMY1F  -72523,-48754</span>
<span class="co">##   [884]     chrY 26206997-26215445      * |            DAZ2         841416</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 26 sequences from an unspecified genome; no seqlengths</span></code></pre>
<p>Being different from the online GREAT, <code><a href="../reference/getRegionGeneAssociations-dispatch.html">getRegionGeneAssociations()</a></code> on the local GREAT object calculates the distance to TSS based on the borders of the input regions. The argument <code>by_middle_points</code> can be set to <code>TRUE</code> to let the distance be based on the middle points of input regions.</p>
<p>Please note the two meta columns are in formats of <code>CharacterList</code> and <code>IntegerList</code> because a region may associate to multiple genes.</p>
<p>Of course you can set a specific geneset term by argument <code>term_id</code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plotRegionGeneAssociations-dispatch.html">plotRegionGeneAssociations</a></span><span class="op">(</span><span class="va">res</span>, term_id <span class="op">=</span> <span class="st">"HALLMARK_APOPTOSIS"</span><span class="op">)</span></code></pre></div>
<p><img src="local-GREAT_files/figure-html/unnamed-chunk-17-1.png" width="960" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/getRegionGeneAssociations-dispatch.html">getRegionGeneAssociations</a></span><span class="op">(</span><span class="va">res</span>, term_id <span class="op">=</span> <span class="st">"HALLMARK_APOPTOSIS"</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## GRanges object with 23 ranges and 2 metadata columns:</span>
<span class="co">##        seqnames            ranges strand | annotated_genes   dist_to_TSS</span>
<span class="co">##           &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;IntegerList&gt;</span>
<span class="co">##    [1]     chr1 40515463-40519356      * |            PPT1         43786</span>
<span class="co">##    [2]     chr1 97386537-97391160      * |            DPYD        995455</span>
<span class="co">##    [3]     chr1 98324960-98333882      * |            DPYD         52733</span>
<span class="co">##    [4]     chr1 98723726-98728003      * |            DPYD       -337111</span>
<span class="co">##    [5]     chr3 18185589-18193193      * |           SATB1        287072</span>
<span class="co">##    ...      ...               ...    ... .             ...           ...</span>
<span class="co">##   [19]    chr19   2644522-2653683      * |         GADD45B        168399</span>
<span class="co">##   [20]    chr20   6656761-6662137      * |            BMP2        -86608</span>
<span class="co">##   [21]    chr20   7709225-7716863      * |            BMP2        960480</span>
<span class="co">##   [22]    chr21 27698626-27707770      * |             APP       -155180</span>
<span class="co">##   [23]    chr22 21271471-21278063      * |           AIFM3        -41355</span>
<span class="co">##   -------</span>
<span class="co">##   seqinfo: 26 sequences from an unspecified genome; no seqlengths</span></code></pre>
</div>
<div class="section level2">
<h2 id="the-shiny-application">The Shiny application<a class="anchor" aria-label="anchor" href="#the-shiny-application"></a>
</h2>
<p><code><a href="../reference/shinyReport-dispatch.html">shinyReport()</a></code> creates a Shiny application to view the complete results:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/shinyReport-dispatch.html">shinyReport</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code><span class="co">## R version 4.1.2 (2021-11-01)</span>
<span class="co">## Platform: x86_64-apple-darwin17.0 (64-bit)</span>
<span class="co">## Running under: macOS Big Sur 10.16</span>
<span class="co">## </span>
<span class="co">## Matrix products: default</span>
<span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib</span>
<span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib</span>
<span class="co">## </span>
<span class="co">## locale:</span>
<span class="co">## [1] C/UTF-8/C/C/C/C</span>
<span class="co">## </span>
<span class="co">## attached base packages:</span>
<span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span>
<span class="co">## [8] base     </span>
<span class="co">## </span>
<span class="co">## other attached packages:</span>
<span class="co">##  [1] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2</span>
<span class="co">##  [2] GenomicFeatures_1.46.3                 </span>
<span class="co">##  [3] AnnotationDbi_1.56.2                   </span>
<span class="co">##  [4] Biobase_2.54.0                         </span>
<span class="co">##  [5] rGREAT_1.99.1                          </span>
<span class="co">##  [6] GenomicRanges_1.46.1                   </span>
<span class="co">##  [7] GenomeInfoDb_1.30.0                    </span>
<span class="co">##  [8] IRanges_2.28.0                         </span>
<span class="co">##  [9] S4Vectors_0.32.3                       </span>
<span class="co">## [10] BiocGenerics_0.40.0                    </span>
<span class="co">## [11] knitr_1.37                             </span>
<span class="co">## </span>
<span class="co">## loaded via a namespace (and not attached):</span>
<span class="co">##   [1] bitops_1.0-7                            </span>
<span class="co">##   [2] matrixStats_0.61.0                      </span>
<span class="co">##   [3] fs_1.5.2                                </span>
<span class="co">##   [4] bit64_4.0.5                             </span>
<span class="co">##   [5] RColorBrewer_1.1-2                      </span>
<span class="co">##   [6] filelock_1.0.2                          </span>
<span class="co">##   [7] progress_1.2.2                          </span>
<span class="co">##   [8] httr_1.4.2                              </span>
<span class="co">##   [9] rprojroot_2.0.2                         </span>
<span class="co">##  [10] tools_4.1.2                             </span>
<span class="co">##  [11] bslib_0.3.1                             </span>
<span class="co">##  [12] utf8_1.2.2                              </span>
<span class="co">##  [13] R6_2.5.1                                </span>
<span class="co">##  [14] DT_0.20                                 </span>
<span class="co">##  [15] colorspace_2.0-2                        </span>
<span class="co">##  [16] DBI_1.1.2                               </span>
<span class="co">##  [17] GetoptLong_1.1.0                        </span>
<span class="co">##  [18] tidyselect_1.1.1                        </span>
<span class="co">##  [19] prettyunits_1.1.1                       </span>
<span class="co">##  [20] bit_4.0.4                               </span>
<span class="co">##  [21] curl_4.3.2                              </span>
<span class="co">##  [22] compiler_4.1.2                          </span>
<span class="co">##  [23] textshaping_0.3.6                       </span>
<span class="co">##  [24] xml2_1.3.3                              </span>
<span class="co">##  [25] desc_1.4.0                              </span>
<span class="co">##  [26] DelayedArray_0.20.0                     </span>
<span class="co">##  [27] rtracklayer_1.54.0                      </span>
<span class="co">##  [28] sass_0.4.0                              </span>
<span class="co">##  [29] rappdirs_0.3.3                          </span>
<span class="co">##  [30] pkgdown_2.0.2                           </span>
<span class="co">##  [31] systemfonts_1.0.3                       </span>
<span class="co">##  [32] stringr_1.4.0                           </span>
<span class="co">##  [33] digest_0.6.29                           </span>
<span class="co">##  [34] Rsamtools_2.10.0                        </span>
<span class="co">##  [35] rmarkdown_2.11                          </span>
<span class="co">##  [36] XVector_0.34.0                          </span>
<span class="co">##  [37] pkgconfig_2.0.3                         </span>
<span class="co">##  [38] htmltools_0.5.2                         </span>
<span class="co">##  [39] MatrixGenerics_1.6.0                    </span>
<span class="co">##  [40] highr_0.9                               </span>
<span class="co">##  [41] dbplyr_2.1.1                            </span>
<span class="co">##  [42] fastmap_1.1.0                           </span>
<span class="co">##  [43] GlobalOptions_0.1.2                     </span>
<span class="co">##  [44] htmlwidgets_1.5.4                       </span>
<span class="co">##  [45] rlang_0.4.12                            </span>
<span class="co">##  [46] RSQLite_2.2.9                           </span>
<span class="co">##  [47] shiny_1.7.1                             </span>
<span class="co">##  [48] shape_1.4.6                             </span>
<span class="co">##  [49] jquerylib_0.1.4                         </span>
<span class="co">##  [50] BiocIO_1.4.0                            </span>
<span class="co">##  [51] generics_0.1.1                          </span>
<span class="co">##  [52] jsonlite_1.7.2                          </span>
<span class="co">##  [53] BiocParallel_1.28.3                     </span>
<span class="co">##  [54] dplyr_1.0.7                             </span>
<span class="co">##  [55] RCurl_1.98-1.5                          </span>
<span class="co">##  [56] magrittr_2.0.1                          </span>
<span class="co">##  [57] GO.db_3.14.0                            </span>
<span class="co">##  [58] GenomeInfoDbData_1.2.7                  </span>
<span class="co">##  [59] Matrix_1.4-0                            </span>
<span class="co">##  [60] Rcpp_1.0.7                              </span>
<span class="co">##  [61] fansi_1.0.2                             </span>
<span class="co">##  [62] lifecycle_1.0.1                         </span>
<span class="co">##  [63] stringi_1.7.6                           </span>
<span class="co">##  [64] yaml_2.2.1                              </span>
<span class="co">##  [65] SummarizedExperiment_1.24.0             </span>
<span class="co">##  [66] zlibbioc_1.40.0                         </span>
<span class="co">##  [67] org.Hs.eg.db_3.14.0                     </span>
<span class="co">##  [68] BiocFileCache_2.2.0                     </span>
<span class="co">##  [69] grid_4.1.2                              </span>
<span class="co">##  [70] blob_1.2.2                              </span>
<span class="co">##  [71] promises_1.2.0.1                        </span>
<span class="co">##  [72] parallel_4.1.2                          </span>
<span class="co">##  [73] crayon_1.4.2                            </span>
<span class="co">##  [74] lattice_0.20-45                         </span>
<span class="co">##  [75] Biostrings_2.62.0                       </span>
<span class="co">##  [76] circlize_0.4.15                         </span>
<span class="co">##  [77] hms_1.1.1                               </span>
<span class="co">##  [78] KEGGREST_1.34.0                         </span>
<span class="co">##  [79] pillar_1.6.4                            </span>
<span class="co">##  [80] rjson_0.2.21                            </span>
<span class="co">##  [81] biomaRt_2.50.2                          </span>
<span class="co">##  [82] XML_3.99-0.8                            </span>
<span class="co">##  [83] glue_1.6.0                              </span>
<span class="co">##  [84] evaluate_0.14                           </span>
<span class="co">##  [85] httpuv_1.6.5                            </span>
<span class="co">##  [86] png_0.1-7                               </span>
<span class="co">##  [87] vctrs_0.3.8                             </span>
<span class="co">##  [88] purrr_0.3.4                             </span>
<span class="co">##  [89] assertthat_0.2.1                        </span>
<span class="co">##  [90] cachem_1.0.6                            </span>
<span class="co">##  [91] xfun_0.29                               </span>
<span class="co">##  [92] mime_0.12                               </span>
<span class="co">##  [93] xtable_1.8-4                            </span>
<span class="co">##  [94] restfulr_0.0.13                         </span>
<span class="co">##  [95] later_1.3.0                             </span>
<span class="co">##  [96] ragg_1.2.1                              </span>
<span class="co">##  [97] TxDb.Hsapiens.UCSC.hg38.knownGene_3.14.0</span>
<span class="co">##  [98] tibble_3.1.6                            </span>
<span class="co">##  [99] GenomicAlignments_1.30.0                </span>
<span class="co">## [100] memoise_2.0.1                           </span>
<span class="co">## [101] ellipsis_0.3.2</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
