<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyze with online GREAT • rGREAT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyze with online GREAT">
<meta property="og:description" content="rGREAT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rGREAT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.1.8</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/local-GREAT.html">Analyze with local GREAT</a>
    </li>
    <li>
      <a href="../articles/online-GREAT.html">Analyze with online GREAT</a>
    </li>
    <li>
      <a href="../articles/other-docs.html">Other documents</a>
    </li>
    <li>
      <a href="../articles/other-geneset-databases.html">Work with other geneset databases</a>
    </li>
    <li>
      <a href="../articles/other-organisms.html">Work with other organisms</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/rGREAT/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyze with online GREAT</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu ( <a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a> )</h4>
            
            <h4 data-toc-skip class="date">2023-04-02</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/rGREAT/blob/HEAD/vignettes/online-GREAT.Rmd" class="external-link"><code>vignettes/online-GREAT.Rmd</code></a></small>
      <div class="hidden name"><code>online-GREAT.Rmd</code></div>

    </div>

    
    
<p><strong>Note: <a href="http://great.stanford.edu/help/display/GREAT/Version+History" class="external-link">On Aug 19 2019 GREAT released version 4</a> which supports <code>hg38</code> genome and removes some ontologies such pathways. <code><a href="../reference/submitGreatJob.html">submitGreatJob()</a></code> still takes <code>hg19</code> as default. <code>hg38</code> can be specified by argument <code>species = "hg38"</code>. To use the older versions such as 3.0.0, specify as <code>submitGreatJob(..., version = "3")</code>.</strong></p>
<p><strong>GREAT</strong> (<a href="http://great.stanford.edu" class="external-link">Genomic Regions Enrichment of Annotations Tool</a>) is a popular web-based tool to associate biological functions to genomic regions. The <strong>rGREAT</strong> package makes GREAT anlaysis automatic by first constructing a HTTP POST request according to user’s input and retrieving results from <strong>GREAT</strong> web server afterwards.</p>
<div class="section level2">
<h2 id="submit-the-job">Submit the job<a class="anchor" aria-label="anchor" href="#submit-the-job"></a>
</h2>
<p>Load the package:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/rGREAT" class="external-link">rGREAT</a></span><span class="op">)</span></span></code></pre></div>
<p>The input data is either a <code>GRanges</code> object or a <em>BED</em>-format data frame, no matter it is sorted or not. In following example, we use a <code>GRanges</code> object which is randomly generated.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">gr</span> <span class="op">=</span> <span class="fu"><a href="../reference/randomRegions.html">randomRegions</a></span><span class="op">(</span>nr <span class="op">=</span> <span class="fl">1000</span>, genome <span class="op">=</span> <span class="st">"hg19"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 0 metadata columns:</span></span>
<span><span class="co">##       seqnames            ranges strand</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span></span>
<span><span class="co">##   [1]     chr1   9204434-9208784      *</span></span>
<span><span class="co">##   [2]     chr1   9853594-9859363      *</span></span>
<span><span class="co">##   [3]     chr1 10862809-10871681      *</span></span>
<span><span class="co">##   [4]     chr1 12716970-12723206      *</span></span>
<span><span class="co">##   [5]     chr1 13814692-13823250      *</span></span>
<span><span class="co">##   [6]     chr1 19243285-19247097      *</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 26 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Submit genomic regions by <code><a href="../reference/submitGreatJob.html">submitGreatJob()</a></code>.</p>
<p>The returned variable <code>job</code> is a <code>GreatJob</code> class instance which can be used to retrieve results from <strong>GREAT</strong> server and store results which are already downloaded.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">job</span> <span class="op">=</span> <span class="fu"><a href="../reference/submitGreatJob.html">submitGreatJob</a></span><span class="op">(</span><span class="va">gr</span><span class="op">)</span></span></code></pre></div>
<p>You can get the summary of your job by directly printing <code>job</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">job</span></span></code></pre></div>
<pre><code><span><span class="co">## Submit time: 2023-04-01 09:44:07 </span></span>
<span><span class="co">##   Note the results may only be avaiable on GREAT server for 24 hours.</span></span>
<span><span class="co">## Version: 4.0.4 </span></span>
<span><span class="co">## Species: hg19 </span></span>
<span><span class="co">## Inputs: 1000 regions</span></span>
<span><span class="co">## Mode: Basal plus extension </span></span>
<span><span class="co">##   Proximal: 5 kb upstream, 1 kb downstream,</span></span>
<span><span class="co">##   plus Distal: up to 1000 kb</span></span>
<span><span class="co">## Include curated regulatory domains</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Enrichment tables for following ontologies have been downloaded:</span></span>
<span><span class="co">##   GO Biological Process</span></span>
<span><span class="co">##   GO Cellular Component</span></span>
<span><span class="co">##   GO Molecular Function</span></span></code></pre>
<p>More parameters can be set for the job:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">job</span> <span class="op">=</span> <span class="fu"><a href="../reference/submitGreatJob.html">submitGreatJob</a></span><span class="op">(</span><span class="va">gr</span>, species <span class="op">=</span> <span class="st">"mm9"</span><span class="op">)</span> <span class="co"># of course, gr should be from mm9</span></span>
<span><span class="va">job</span> <span class="op">=</span> <span class="fu"><a href="../reference/submitGreatJob.html">submitGreatJob</a></span><span class="op">(</span><span class="va">gr</span>, adv_upstream <span class="op">=</span> <span class="fl">10</span>, adv_downstream <span class="op">=</span> <span class="fl">2</span>, adv_span <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">job</span> <span class="op">=</span> <span class="fu"><a href="../reference/submitGreatJob.html">submitGreatJob</a></span><span class="op">(</span><span class="va">gr</span>, rule <span class="op">=</span> <span class="st">"twoClosest"</span>, adv_twoDistance <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="va">job</span> <span class="op">=</span> <span class="fu"><a href="../reference/submitGreatJob.html">submitGreatJob</a></span><span class="op">(</span><span class="va">gr</span>, rule <span class="op">=</span> <span class="st">"oneClosest"</span>, adv_oneDistance <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span></code></pre></div>
<p>Also you can choose different versions of GREAT for the analysis.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">job</span> <span class="op">=</span> <span class="fu"><a href="../reference/submitGreatJob.html">submitGreatJob</a></span><span class="op">(</span><span class="va">gr</span>, version <span class="op">=</span> <span class="st">"3.0"</span><span class="op">)</span></span>
<span><span class="va">job</span> <span class="op">=</span> <span class="fu"><a href="../reference/submitGreatJob.html">submitGreatJob</a></span><span class="op">(</span><span class="va">gr</span>, version <span class="op">=</span> <span class="st">"2.0"</span><span class="op">)</span></span></code></pre></div>
<p style="color:red">
Note: from <b>rGREAT package</b> 1.99.0, background by <code>bg</code> argument is not supported any more (currently you can still use it, but you will see a warning message), because <b>GREAT</b> requires a special format for <code>gr</code> and <code>bg</code> if both are set, and it uses a different method for the enrichment analysis and returns enrichment tables in a different format. <a href="local-GREAT.html#Set_background_regions">But still, you can use local <b>GREAT</b> to integrate background regions.</a> Seel the rGREAT paper for more details.
</p>
<p>Available parameters are (following content is copied from <strong>GREAT</strong> website):</p>
<ul>
<li>
<code>species</code>: “hg38”, “hg19”, “mm10”, “mm9” are supported in GREAT version 4.x.x, “hg19”, “mm10”, “mm9”, “danRer7” are supported in GREAT version 3.x.x and “hg19”, “hg18”, “mm9”, “danRer7” are supported in GREAT version 2.x.x.</li>
<li>
<code>includeCuratedRegDoms</code>: Whether to include curated regulatory domains.</li>
<li>
<code>rule</code>: How to associate genomic regions to genes.
<ul>
<li>
<code>basalPlusExt</code>: mode ‘Basal plus extension’. Gene regulatory domain definition: Each gene is assigned a basal regulatory domain of a minimum distance upstream and downstream of the TSS (regardless of other nearby genes). The gene regulatory domain is extended in both directions to the nearest gene’s basal domain but no more than the maximum extension in one direction.
<ul>
<li>
<code>adv_upstream</code>: proximal extension to upstream (unit: kb)</li>
<li>
<code>adv_downstream</code>: proximal extension to downstream (unit: kb)</li>
<li>
<code>adv_span</code>: maximum extension (unit: kb)</li>
</ul>
</li>
<li>
<code>twoClosest</code>: mode ‘Two nearest genes’. Gene regulatory domain definition: Each gene is assigned a regulatory domain that extends in both directions to the nearest gene’s TSS but no more than the maximum extension in one direction.
<ul>
<li>
<code>adv_twoDistance</code>: maximum extension (unit: kb)</li>
</ul>
</li>
<li>
<code>oneClosest</code>: mode ‘Single nearest gene’. Gene regulatory domain definition: Each gene is assigned a regulatory domain that extends in both directions to the midpoint between the gene’s TSS and the nearest gene’s TSS but no more than the maximum extension in one direction.
<ul>
<li>
<code>adv_oneDistance</code>: maximum extension (unit: kb)</li>
</ul>
</li>
</ul>
</li>
</ul>
<p><strong>GREAT</strong> uses <a href="https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655452/File+Formats#FileFormats-WhatisBEDformat?" class="external-link">the UCSC bed-format</a> where genomic coordinates are 0-based. Many R packages generate genomic regions as 1-based. Thus by default, the start positions of regions are subtracted by 1. If your regions are already 0-based, you can specify <code>gr_is_zero_based = TRUE</code> in <code><a href="../reference/submitGreatJob.html">submitGreatJob()</a></code>. Anyway in most cases, this will only slightly affect the enrichment results.</p>
</div>
<div class="section level2">
<h2 id="get-enrichment-tables">Get enrichment tables<a class="anchor" aria-label="anchor" href="#get-enrichment-tables"></a>
</h2>
<p>With <code>job</code>, we can now retrieve results from <strong>GREAT</strong>. The first and the primary results are the tables which contain enrichment statistics for the analysis. By default it will retrieve results from three GO Ontologies. All tables contains statistics for all terms no matter they are significant or not. Users can then make filtering with self-defined cutoff.</p>
<p>There is a column for adjusted p-values by “BH” method. Other p-value adjustment methods can be applied by <code><a href="https://rdrr.io/r/stats/p.adjust.html" class="external-link">p.adjust()</a></code>.</p>
<p>The returned value of <code><a href="../reference/getEnrichmentTables-dispatch.html">getEnrichmentTables()</a></code> is a list of data frames in which each one corresponds to the table for a single ontology. The structure of data frames are same as the tables on <strong>GREAT</strong> website.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tbl</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTables-dispatch.html">getEnrichmentTables</a></span><span class="op">(</span><span class="va">job</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The default enrichment table does not contain informatin of associated</span></span>
<span><span class="co">## genes for each input region. You can set `download_by = 'tsv'` to</span></span>
<span><span class="co">## download the complete table, but note only the top 500 regions can be</span></span>
<span><span class="co">## retreived. See the following link:</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655401/Export#Export-GlobalExport</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Except the additional gene-region association column if taking 'tsv' as</span></span>
<span><span class="co">## the source of result, all other columns are the same if you choose</span></span>
<span><span class="co">## 'json' (the default) as the source. Or you can try the local GREAT</span></span>
<span><span class="co">## analysis with the function `great()`.</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">tbl</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GO Molecular Function" "GO Biological Process" "GO Cellular Component"</span></span></code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tbl</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, <span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##           ID                                                           name</span></span>
<span><span class="co">## 1 GO:0070696 transmembrane receptor protein serine/threonine kinase binding</span></span>
<span><span class="co">## 2 GO:0033612                       receptor serine/threonine kinase binding</span></span>
<span><span class="co">##   Binom_Genome_Fraction Binom_Expected Binom_Observed_Region_Hits</span></span>
<span><span class="co">## 1           0.003455733       3.455733                         11</span></span>
<span><span class="co">## 2           0.003596413       3.596413                         11</span></span>
<span><span class="co">##   Binom_Fold_Enrichment Binom_Region_Set_Coverage Binom_Raw_PValue</span></span>
<span><span class="co">## 1              3.183116                     0.011     0.0008981541</span></span>
<span><span class="co">## 2              3.058603                     0.011     0.0012301660</span></span>
<span><span class="co">##   Binom_Adjp_BH Hyper_Total_Genes Hyper_Expected Hyper_Observed_Gene_Hits</span></span>
<span><span class="co">## 1             1                13      0.9938002                        5</span></span>
<span><span class="co">## 2             1                15      1.1466930                        5</span></span>
<span><span class="co">##   Hyper_Fold_Enrichment Hyper_Gene_Set_Coverage Hyper_Term_Gene_Coverage</span></span>
<span><span class="co">## 1              5.031192             0.003526093                0.3846154</span></span>
<span><span class="co">## 2              4.360367             0.003526093                0.3333333</span></span>
<span><span class="co">##   Hyper_Raw_PValue Hyper_Adjp_BH</span></span>
<span><span class="co">## 1      0.001982437     0.4919942</span></span>
<span><span class="co">## 2      0.004066220     0.6862153</span></span></code></pre>
<p>Information stored in <code>job</code> will be updated after retrieving enrichment tables.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">job</span></span></code></pre></div>
<pre><code><span><span class="co">## Submit time: 2023-04-01 09:44:07 </span></span>
<span><span class="co">##   Note the results may only be avaiable on GREAT server for 24 hours.</span></span>
<span><span class="co">## Version: 4.0.4 </span></span>
<span><span class="co">## Species: hg19 </span></span>
<span><span class="co">## Inputs: 1000 regions</span></span>
<span><span class="co">## Mode: Basal plus extension </span></span>
<span><span class="co">##   Proximal: 5 kb upstream, 1 kb downstream,</span></span>
<span><span class="co">##   plus Distal: up to 1000 kb</span></span>
<span><span class="co">## Include curated regulatory domains</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Enrichment tables for following ontologies have been downloaded:</span></span>
<span><span class="co">##   GO Biological Process</span></span>
<span><span class="co">##   GO Cellular Component</span></span>
<span><span class="co">##   GO Molecular Function</span></span></code></pre>
<p>You can get results by either specifying the ontologies or by the pre-defined categories (categories already contains pre-defined sets of ontologies):</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tbl</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTables-dispatch.html">getEnrichmentTables</a></span><span class="op">(</span><span class="va">job</span>, ontology <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO Molecular Function"</span>, <span class="st">"Human Phenotype"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">tbl</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTables-dispatch.html">getEnrichmentTables</a></span><span class="op">(</span><span class="va">job</span>, category <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GO"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>As you have seen in the previous messages and results, The enrichment tables contain no associated genes. However, you can set <code>download_by = 'tsv'</code> in <code><a href="../reference/getEnrichmentTables-dispatch.html">getEnrichmentTables()</a></code> to download the complete tables, but due to <a href="https://great-help.atlassian.net/wiki/spaces/GREAT/pages/655401/Export#Export-GlobalExport" class="external-link">the restriction from GREAT web server</a>, only the top 500 regions can be retreived (check the last two columns of <code>tbl2[["GO Molecular Function"]]</code> in the following example).</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tbl2</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTables-dispatch.html">getEnrichmentTables</a></span><span class="op">(</span><span class="va">job</span>, download_by <span class="op">=</span> <span class="st">"tsv"</span><span class="op">)</span></span></code></pre></div>
<p>All available ontology names for a given species can be get by <code><a href="../reference/availableOntologies-GreatJob-method.html">availableOntologies()</a></code> and all available ontology categories can be get by <code><a href="../reference/availableCategories-GreatJob-method.html">availableCategories()</a></code>. Here you do not need to provide species information because <code>job</code> already contains it.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/availableOntologies-GreatJob-method.html">availableOntologies</a></span><span class="op">(</span><span class="va">job</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GO Molecular Function"     "GO Biological Process"    </span></span>
<span><span class="co">## [3] "GO Cellular Component"     "Mouse Phenotype"          </span></span>
<span><span class="co">## [5] "Mouse Phenotype Single KO" "Human Phenotype"          </span></span>
<span><span class="co">## [7] "Ensembl Genes"</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/availableCategories-GreatJob-method.html">availableCategories</a></span><span class="op">(</span><span class="va">job</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GO"        "Phenotype" "Genes"</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/availableOntologies-GreatJob-method.html">availableOntologies</a></span><span class="op">(</span><span class="va">job</span>, category <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "GO Molecular Function" "GO Biological Process" "GO Cellular Component"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="make-volcano-plot">Make volcano plot<a class="anchor" aria-label="anchor" href="#make-volcano-plot"></a>
</h2>
<p>In differential gene expression analysis, volcano plot is used to visualize relations between log2 fold change and (adjusted) p-values. Similarly, we can also use volcano plot to visualize relations between fold enrichment and (adjusted) p-values for the enrichment analysis. The plot is made by the function <code><a href="../reference/plotVolcano-dispatch.html">plotVolcano()</a></code>:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotVolcano-dispatch.html">plotVolcano</a></span><span class="op">(</span><span class="va">job</span>, ontology <span class="op">=</span> <span class="st">"GO Biological Process"</span><span class="op">)</span></span></code></pre></div>
<p><img src="online-GREAT_files/figure-html/unnamed-chunk-14-1.png" width="576" style="display: block; margin: auto;"></p>
<p>As the enrichment analysis basically only looks for over-representations, it is actually half volcano.</p>
</div>
<div class="section level2">
<h2 id="get-region-gene-associations">Get region-gene associations<a class="anchor" aria-label="anchor" href="#get-region-gene-associations"></a>
</h2>
<p>Association between genomic regions and genes can be plotted by <code><a href="../reference/plotRegionGeneAssociations-dispatch.html">plotRegionGeneAssociations()</a></code>. The function will make the three plots which are same as on <strong>GREAT</strong> website.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegionGeneAssociations-dispatch.html">plotRegionGeneAssociations</a></span><span class="op">(</span><span class="va">job</span><span class="op">)</span></span></code></pre></div>
<p><img src="online-GREAT_files/figure-html/unnamed-chunk-15-1.png" width="960" style="display: block; margin: auto;"></p>
<p><code><a href="../reference/getRegionGeneAssociations-dispatch.html">getRegionGeneAssociations()</a></code> returns a <code>GRanges</code> object which contains the gene-region associations. <span style="color:red;">Note the column <code>dist_to_TSS</code> is based on the middle points of the input regions to TSS.</span></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getRegionGeneAssociations-dispatch.html">getRegionGeneAssociations</a></span><span class="op">(</span><span class="va">job</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 885 ranges and 2 metadata columns:</span></span>
<span><span class="co">##         seqnames            ranges strand | annotated_genes    dist_to_TSS</span></span>
<span><span class="co">##            &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt;  &lt;IntegerList&gt;</span></span>
<span><span class="co">##     [1]     chr1   9204434-9208784      * |     H6PD,GPR157  -88225,-17380</span></span>
<span><span class="co">##     [2]     chr1   9853594-9859363      * |   CLSTN1,PIK3CD   28106,144675</span></span>
<span><span class="co">##     [3]     chr1 10862809-10871681      * |    TARDBP,CASZ1 -205454,-10540</span></span>
<span><span class="co">##     [4]     chr1 12716970-12723206      * | AADACL3,AADACL4   -56058,15522</span></span>
<span><span class="co">##     [5]     chr1 13814692-13823250      * | LRRC38,PRAMEF20    21572,82064</span></span>
<span><span class="co">##     ...      ...               ...    ... .             ...            ...</span></span>
<span><span class="co">##   [881]     chrY 16384799-16389852      * |    NLGN4Y,VCY1B -248301,219228</span></span>
<span><span class="co">##   [882]     chrY 16842593-16846893      * |          NLGN4Y         209117</span></span>
<span><span class="co">##   [883]     chrY 16998750-17003225      * |          NLGN4Y         365361</span></span>
<span><span class="co">##   [884]     chrY 24401612-24406252      * |   RBMY1J,RBMY1F -145685,-74837</span></span>
<span><span class="co">##   [885]     chrY 26206997-26215445      * |     BPY2B,CDY1B -552930,-17105</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 24 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<p>Please note the two meta columns are in formats of <code>CharacterList</code> and <code>IntegerList</code> because a region may associate to multiple genes.</p>
<p>You can also choose only plotting one of the three figures.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegionGeneAssociations-dispatch.html">plotRegionGeneAssociations</a></span><span class="op">(</span><span class="va">job</span>, which_plot <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<p>By specifying ontology and term ID, you can get the associations in a certain term. Here the term ID is from the first column of the data frame from <code><a href="../reference/getEnrichmentTables-dispatch.html">getEnrichmentTables()</a></code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotRegionGeneAssociations-dispatch.html">plotRegionGeneAssociations</a></span><span class="op">(</span><span class="va">job</span>, ontology <span class="op">=</span> <span class="st">"GO Molecular Function"</span>,</span>
<span>    term_id <span class="op">=</span> <span class="st">"GO:0004984"</span><span class="op">)</span></span></code></pre></div>
<p><img src="online-GREAT_files/figure-html/unnamed-chunk-18-1.png" width="960" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/getRegionGeneAssociations-dispatch.html">getRegionGeneAssociations</a></span><span class="op">(</span><span class="va">job</span>, ontology <span class="op">=</span> <span class="st">"GO Molecular Function"</span>,</span>
<span>    term_id <span class="op">=</span> <span class="st">"GO:0004984"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 12 ranges and 2 metadata columns:</span></span>
<span><span class="co">##        seqnames              ranges strand | annotated_genes   dist_to_TSS</span></span>
<span><span class="co">##           &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;IntegerList&gt;</span></span>
<span><span class="co">##    [1]     chr1 247871555-247874068      * |    OR13G1,OR6F1   -36446,3294</span></span>
<span><span class="co">##    [2]     chr5 180577320-180586838      * |           OR2V2           136</span></span>
<span><span class="co">##    [3]    chr11     4669904-4675668      * |   OR51E1,OR51E2    8136,46286</span></span>
<span><span class="co">##    [4]    chr11   50163533-50171101      * |          OR4C12       -163246</span></span>
<span><span class="co">##    [5]    chr11   51961271-51969107      * |          OR4C46        449907</span></span>
<span><span class="co">##    ...      ...                 ...    ... .             ...           ...</span></span>
<span><span class="co">##    [8]    chr12   48643533-48645814      * |         OR10AD1        -47503</span></span>
<span><span class="co">##    [9]    chr14   18596966-18604599      * |         OR11H12       -776740</span></span>
<span><span class="co">##   [10]    chr14   20760309-20769811      * |          OR11H4         54162</span></span>
<span><span class="co">##   [11]    chr14   22807516-22814493      * |           OR4E2        677707</span></span>
<span><span class="co">##   [12]    chr22   16218127-16222924      * |          OR11H1        229280</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 6 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="the-shiny-application">The Shiny application<a class="anchor" aria-label="anchor" href="#the-shiny-application"></a>
</h2>
<p><code><a href="../reference/shinyReport-dispatch.html">shinyReport()</a></code> creates a Shiny application to view the complete results:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/shinyReport-dispatch.html">shinyReport</a></span><span class="op">(</span><span class="va">job</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.0 (2022-04-22)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin17.0 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Big Sur/Monterey 10.16</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib</span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] C/UTF-8/C/C/C/C</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] rGREAT_2.1.8         GenomicRanges_1.50.2 GenomeInfoDb_1.34.9 </span></span>
<span><span class="co">## [4] IRanges_2.32.0       S4Vectors_0.36.1     BiocGenerics_0.44.0 </span></span>
<span><span class="co">## [7] knitr_1.42          </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] bitops_1.0-7                            </span></span>
<span><span class="co">##   [2] matrixStats_0.63.0                      </span></span>
<span><span class="co">##   [3] fs_1.6.1                                </span></span>
<span><span class="co">##   [4] bit64_4.0.5                             </span></span>
<span><span class="co">##   [5] doParallel_1.0.17                       </span></span>
<span><span class="co">##   [6] RColorBrewer_1.1-3                      </span></span>
<span><span class="co">##   [7] filelock_1.0.2                          </span></span>
<span><span class="co">##   [8] progress_1.2.2                          </span></span>
<span><span class="co">##   [9] httr_1.4.4                              </span></span>
<span><span class="co">##  [10] rprojroot_2.0.3                         </span></span>
<span><span class="co">##  [11] tools_4.2.0                             </span></span>
<span><span class="co">##  [12] bslib_0.4.2                             </span></span>
<span><span class="co">##  [13] utf8_1.2.3                              </span></span>
<span><span class="co">##  [14] R6_2.5.1                                </span></span>
<span><span class="co">##  [15] DT_0.27                                 </span></span>
<span><span class="co">##  [16] colorspace_2.1-0                        </span></span>
<span><span class="co">##  [17] DBI_1.1.3                               </span></span>
<span><span class="co">##  [18] GetoptLong_1.1.0                        </span></span>
<span><span class="co">##  [19] tidyselect_1.2.0                        </span></span>
<span><span class="co">##  [20] prettyunits_1.1.1                       </span></span>
<span><span class="co">##  [21] bit_4.0.5                               </span></span>
<span><span class="co">##  [22] curl_5.0.0                              </span></span>
<span><span class="co">##  [23] compiler_4.2.0                          </span></span>
<span><span class="co">##  [24] textshaping_0.3.6                       </span></span>
<span><span class="co">##  [25] cli_3.6.0                               </span></span>
<span><span class="co">##  [26] Biobase_2.58.0                          </span></span>
<span><span class="co">##  [27] xml2_1.3.3                              </span></span>
<span><span class="co">##  [28] desc_1.4.2                              </span></span>
<span><span class="co">##  [29] DelayedArray_0.24.0                     </span></span>
<span><span class="co">##  [30] rtracklayer_1.58.0                      </span></span>
<span><span class="co">##  [31] sass_0.4.5                              </span></span>
<span><span class="co">##  [32] rappdirs_0.3.3                          </span></span>
<span><span class="co">##  [33] pkgdown_2.0.7                           </span></span>
<span><span class="co">##  [34] systemfonts_1.0.4                       </span></span>
<span><span class="co">##  [35] stringr_1.5.0                           </span></span>
<span><span class="co">##  [36] digest_0.6.31                           </span></span>
<span><span class="co">##  [37] Rsamtools_2.14.0                        </span></span>
<span><span class="co">##  [38] rmarkdown_2.20                          </span></span>
<span><span class="co">##  [39] XVector_0.38.0                          </span></span>
<span><span class="co">##  [40] pkgconfig_2.0.3                         </span></span>
<span><span class="co">##  [41] htmltools_0.5.4                         </span></span>
<span><span class="co">##  [42] MatrixGenerics_1.10.0                   </span></span>
<span><span class="co">##  [43] highr_0.10                              </span></span>
<span><span class="co">##  [44] dbplyr_2.3.0                            </span></span>
<span><span class="co">##  [45] fastmap_1.1.0                           </span></span>
<span><span class="co">##  [46] GlobalOptions_0.1.2                     </span></span>
<span><span class="co">##  [47] htmlwidgets_1.6.1                       </span></span>
<span><span class="co">##  [48] rlang_1.0.6                             </span></span>
<span><span class="co">##  [49] RSQLite_2.2.20                          </span></span>
<span><span class="co">##  [50] shiny_1.7.4                             </span></span>
<span><span class="co">##  [51] shape_1.4.6                             </span></span>
<span><span class="co">##  [52] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2 </span></span>
<span><span class="co">##  [53] jquerylib_0.1.4                         </span></span>
<span><span class="co">##  [54] BiocIO_1.8.0                            </span></span>
<span><span class="co">##  [55] generics_0.1.3                          </span></span>
<span><span class="co">##  [56] jsonlite_1.8.4                          </span></span>
<span><span class="co">##  [57] BiocParallel_1.32.5                     </span></span>
<span><span class="co">##  [58] dplyr_1.1.0                             </span></span>
<span><span class="co">##  [59] RCurl_1.98-1.10                         </span></span>
<span><span class="co">##  [60] magrittr_2.0.3                          </span></span>
<span><span class="co">##  [61] GO.db_3.16.0                            </span></span>
<span><span class="co">##  [62] GenomeInfoDbData_1.2.9                  </span></span>
<span><span class="co">##  [63] Matrix_1.5-3                            </span></span>
<span><span class="co">##  [64] Rcpp_1.0.10                             </span></span>
<span><span class="co">##  [65] fansi_1.0.4                             </span></span>
<span><span class="co">##  [66] lifecycle_1.0.3                         </span></span>
<span><span class="co">##  [67] stringi_1.7.12                          </span></span>
<span><span class="co">##  [68] yaml_2.3.7                              </span></span>
<span><span class="co">##  [69] SummarizedExperiment_1.28.0             </span></span>
<span><span class="co">##  [70] zlibbioc_1.44.0                         </span></span>
<span><span class="co">##  [71] org.Hs.eg.db_3.16.0                     </span></span>
<span><span class="co">##  [72] BiocFileCache_2.6.0                     </span></span>
<span><span class="co">##  [73] grid_4.2.0                              </span></span>
<span><span class="co">##  [74] blob_1.2.3                              </span></span>
<span><span class="co">##  [75] promises_1.2.0.1                        </span></span>
<span><span class="co">##  [76] parallel_4.2.0                          </span></span>
<span><span class="co">##  [77] crayon_1.5.2                            </span></span>
<span><span class="co">##  [78] lattice_0.20-45                         </span></span>
<span><span class="co">##  [79] Biostrings_2.66.0                       </span></span>
<span><span class="co">##  [80] GenomicFeatures_1.50.4                  </span></span>
<span><span class="co">##  [81] circlize_0.4.16                         </span></span>
<span><span class="co">##  [82] hms_1.1.2                               </span></span>
<span><span class="co">##  [83] KEGGREST_1.38.0                         </span></span>
<span><span class="co">##  [84] pillar_1.8.1                            </span></span>
<span><span class="co">##  [85] rjson_0.2.21                            </span></span>
<span><span class="co">##  [86] codetools_0.2-19                        </span></span>
<span><span class="co">##  [87] biomaRt_2.54.0                          </span></span>
<span><span class="co">##  [88] XML_3.99-0.13                           </span></span>
<span><span class="co">##  [89] glue_1.6.2                              </span></span>
<span><span class="co">##  [90] evaluate_0.20                           </span></span>
<span><span class="co">##  [91] httpuv_1.6.8                            </span></span>
<span><span class="co">##  [92] foreach_1.5.2                           </span></span>
<span><span class="co">##  [93] png_0.1-8                               </span></span>
<span><span class="co">##  [94] vctrs_0.5.2                             </span></span>
<span><span class="co">##  [95] purrr_1.0.1                             </span></span>
<span><span class="co">##  [96] assertthat_0.2.1                        </span></span>
<span><span class="co">##  [97] cachem_1.0.6                            </span></span>
<span><span class="co">##  [98] xfun_0.37                               </span></span>
<span><span class="co">##  [99] mime_0.12                               </span></span>
<span><span class="co">## [100] xtable_1.8-4                            </span></span>
<span><span class="co">## [101] restfulr_0.0.15                         </span></span>
<span><span class="co">## [102] later_1.3.0                             </span></span>
<span><span class="co">## [103] ragg_1.2.5                              </span></span>
<span><span class="co">## [104] TxDb.Hsapiens.UCSC.hg38.knownGene_3.16.0</span></span>
<span><span class="co">## [105] tibble_3.1.8                            </span></span>
<span><span class="co">## [106] iterators_1.0.14                        </span></span>
<span><span class="co">## [107] GenomicAlignments_1.34.0                </span></span>
<span><span class="co">## [108] AnnotationDbi_1.60.0                    </span></span>
<span><span class="co">## [109] memoise_2.0.1                           </span></span>
<span><span class="co">## [110] ellipsis_0.3.2</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
