<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Suppl 1. Use background regions • rGREAT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Suppl 1. Use background regions">
<meta property="og:description" content="rGREAT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rGREAT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Online GREAT</li>
    <li>
      <a href="../articles/online-GREAT.html">Analyze with online GREAT</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Local GREAT</li>
    <li>
      <a href="../articles/local-GREAT.html">Analyze with local GREAT</a>
    </li>
    <li>
      <a href="../articles/other-geneset-databases.html">Work with other geneset databases</a>
    </li>
    <li>
      <a href="../articles/other-organisms.html">Work with other organisms</a>
    </li>
    <li>
      <a href="../articles/group_associations.html">Get gene-region associations for a group of GO terms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Supplementaries in the rGREAT paper</li>
    <li>
      <a href="../articles/suppl_use_background.html">Suppl 1. Use background regions</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_GO.html">Suppl 2. Compare GO annotations from online and local GREAT</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_online_and_local.html">Suppl 3. Compare online and local GREAT</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_tss.html">Suppl 4. Compare different TSS sources</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_genome_version.html">Suppl 5. Compare different genome versions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/rGREAT/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Suppl 1. Use background regions</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu ( <a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a> )</h4>
            
            <h4 data-toc-skip class="date">2024-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/rGREAT/blob/HEAD/vignettes/suppl_use_background.Rmd" class="external-link"><code>vignettes/suppl_use_background.Rmd</code></a></small>
      <div class="hidden name"><code>suppl_use_background.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="a-general-example">A general example<a class="anchor" aria-label="anchor" href="#a-general-example"></a>
</h2>
<p>In this document, we will discuss the use of background regions. We first demonstrate it with a ChIP-seq TFBS dataset from <a href="https://genome.ucsc.edu/cgi-bin/hgTables" class="external-link">UCSC table browser</a>. Parameters are:</p>
<p>In the “Select dataset” section:</p>
<pre><code>clade = Mammal
genome = Human
assembly = GRCh37/hg19
group = Regulation
track = ENCODE 3 TFBS
table: GM12878 MYB</code></pre>
<p>And in the “Retrieve and display data” section:</p>
<pre><code>output format = BED - browser extensible data</code></pre>
<p>Then click the button “get output”.</p>
<p>We first read it as a <code>GRanges</code> object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/rGREAT" class="external-link">rGREAT</a></span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"data/tb_encTfChipPkENCFF215YWS_GM12878_MYB_hg19.bed"</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">=</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">gr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="va">df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1</span>, <span class="va">df</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The next two GREAT analysis uses the whole genome as background and excludes gap regions.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res1</span> <span class="op">=</span> <span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"hg19"</span>, exclude <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">res2</span> <span class="op">=</span> <span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"hg19"</span>, exclude <span class="op">=</span> <span class="st">"gap"</span><span class="op">)</span></span></code></pre></div>
<p>And we compare the significant GO terms:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tb1</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable</a></span><span class="op">(</span><span class="va">res1</span><span class="op">)</span></span>
<span><span class="va">tb2</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable</a></span><span class="op">(</span><span class="va">res2</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jolars/eulerr" class="external-link">eulerr</a></span><span class="op">)</span></span>
<span><span class="va">lt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    genome <span class="op">=</span> <span class="va">tb1</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">tb1</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.001</span><span class="op">]</span>,</span>
<span>    exclude_gap <span class="op">=</span> <span class="va">tb2</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">tb2</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.001</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://jolars.github.io/eulerr/reference/euler.html" class="external-link">euler</a></span><span class="op">(</span><span class="va">lt</span><span class="op">)</span>, quantities <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_use_background_files/figure-html/unnamed-chunk-4-1.png" width="700" style="display: block; margin: auto;"></p>
<p>We can see, when excluding gap regions, there are fewer significant terms left. In GREAT analysis where Binomial test is applied, denote following variables: <span class="math inline">\(N\)</span> as total number of input regions, <span class="math inline">\(n\)</span> as number of input regions that fall into the extended TSS regions (denote the corresponding random variable as <span class="math inline">\(X\)</span>), and <span class="math inline">\(p\)</span> as the fraction of extended TSS regions in the genome, then <span class="math inline">\(X \sim B(p, N)\)</span>.</p>
<p>When gap regions are removed from the analysis, <span class="math inline">\(N\)</span> and <span class="math inline">\(n\)</span> are most likely unchanged because the input regions will not overlap to the gap regions, but <span class="math inline">\(p\)</span> gets higher because the coverage of extended TSS regions is most likely unchanged, but the background size (genome subtracting gap regions) becomes smaller. When <span class="math inline">\(p\)</span> gets higher while <span class="math inline">\(N\)</span> and <span class="math inline">\(n\)</span> are unchanged, the p-value from Binomial test also gets higher (when <span class="math inline">\(p\)</span> gets higher, the probability of obtaining <span class="math inline">\(n\)</span> regions from <span class="math inline">\(N\)</span> with a success rate of <span class="math inline">\(p\)</span> increases), which decreases the number of significant terms.</p>
<p>The null assumption of the Binomial test is that input regions are uniformaly distributed in the genome. Since gap regions are not sequenced and input regions will never overlap to them, using whole genome as background will not be proper and it under-estimates the fraction of a function associated regions in the genome, i.e. the value of <span class="math inline">\(p\)</span>, which results more smaller p-values and possible more false positives.</p>
<p>This might happen for other scenations, e.g. when dealing with methylation-related regions (e.g. differentially methylated regions, DMRs), choosing a background showing similar CpG density might be more proper, which can decrease the false positives caused by the regions that will never be called as DMRs (e.g. CpG poor regions).</p>
</div>
<div class="section level2">
<h2 id="set-background-by-chromosomes">Set background by chromosomes<a class="anchor" aria-label="anchor" href="#set-background-by-chromosomes"></a>
</h2>
<p>This scenario might be less used, but it is a good example to show the idea. In <code><a href="../reference/great.html">great()</a></code>, arguments <code>background</code> and <code>exclude</code> can be set to a vector of chromosomes to keep or to remove certain chromosomes, then all analysis is restricted in the selected chromosomes. In the following example, we apply GREAT analysis only on one single chromosome at a time.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">chr</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">res_list</span><span class="op">[[</span><span class="va">chr</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"hg19"</span>, background <span class="op">=</span> <span class="va">chr</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>Next we compare the signigicant GO terms. In the followig plot, each row is a GO term, the first heatmap shows whether the GO term is significant (red) in the corresponding chromosome and the right square heatmap shows the similarities of GO terms. GO terms are clustered into several groups by clustering the similarity matrix. Summaries of GO terms in clusters are represented as word clouds and they are attached to the heatmap.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tb</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">tb</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">tb</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/simplifyEnrichment" class="external-link">simplifyEnrichment</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/simplifyEnrichment/man/simplifyGOFromMultipleLists.html" class="external-link">simplifyGOFromMultipleLists</a></span><span class="op">(</span><span class="va">sig_list</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_use_background_files/figure-html/unnamed-chunk-6-1.png" width="1344" style="display: block; margin: auto;"></p>
<p>From the plot we can see there are some degrees of specificities of the enrichment on different chromosomes, which are mainly caused by the unequal distribution of genes on chromosomes in gene sets.</p>
</div>
<div class="section level2">
<h2 id="set-background-as-a-set-of-genomic-regions">Set background as a set of genomic regions<a class="anchor" aria-label="anchor" href="#set-background-as-a-set-of-genomic-regions"></a>
</h2>
<p>It is more common to set background by a set of pre-defined genomic regions. In the following example, we use the chromatin states dataset from the same cell line as in the first example. Parameters for retrieving the data from UCSC table browser are:</p>
<pre><code>clade = Mammal
genome = Human
assembly = GRCh37/hg19
group = Regulation
track = Broad ChromHMM
table: GM12878 ChromHMM</code></pre>
<p>We read the chromatin states data and format it as a list where each element in the list corresponds to regions in one single chromatin state.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"data/GM12878_chromHMM.bed.gz"</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">=</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">all_states</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="va">df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">+</span> <span class="fl">1</span>, <span class="va">df</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span>, </span>
<span>    state <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"^\\d+_"</span>, <span class="st">""</span>, <span class="va">df</span><span class="op">[</span>, <span class="fl">4</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">all_states</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/splitAsList.html" class="external-link">split</a></span><span class="op">(</span><span class="va">all_states</span>, <span class="va">all_states</span><span class="op">$</span><span class="va">state</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">all_states</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Active_Promoter" "Heterochrom/lo"  "Insulator"       "Poised_Promoter"</span></span>
<span><span class="co">##  [5] "Repetitive/CNV"  "Repressed"       "Strong_Enhancer" "Txn_Elongation" </span></span>
<span><span class="co">##  [9] "Txn_Transition"  "Weak_Enhancer"   "Weak_Promoter"   "Weak_Txn"</span></span></code></pre>
<p>Each time, we set background to only one chromatin state:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">all_states</span>, <span class="kw">function</span><span class="op">(</span><span class="va">bg</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"hg19"</span>, background <span class="op">=</span> <span class="va">bg</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre></div>
<p>We compare the numbers of significant GO terms:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tb</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">tb</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">tb</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">sig_list</span>, <span class="va">length</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Active_Promoter  Heterochrom/lo       Insulator Poised_Promoter  Repetitive/CNV </span></span>
<span><span class="co">##              35             310               2               0               0 </span></span>
<span><span class="co">##       Repressed Strong_Enhancer  Txn_Elongation  Txn_Transition   Weak_Enhancer </span></span>
<span><span class="co">##               0             208               0               0              32 </span></span>
<span><span class="co">##   Weak_Promoter        Weak_Txn </span></span>
<span><span class="co">##              32               8</span></span></code></pre>
<p>It is interesting to see there are a lot of GO terms enriched when using heterchromatin as background. In the following parts of this document, we only look at three chromatin states: active promoter, strong enhancer and heterochromatin:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_list</span> <span class="op">=</span> <span class="va">res_list</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Active_Promoter"</span>, <span class="st">"Strong_Enhancer"</span>, <span class="st">"Heterochrom/lo"</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<p>Next two plots demonstrate total widths of background regions in different states, and numbers of TFBS peaks falling in different background.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">8</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span>, mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, xpd <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="va">w</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/start.html" class="external-link">width</a></span><span class="op">(</span><span class="va">x</span><span class="op">@</span><span class="va">background</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="va">w</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span>, las <span class="op">=</span> <span class="fl">1</span>, xlab <span class="op">=</span> <span class="st">"Base pairs"</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2e9</span><span class="op">)</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Sum of widths / background width"</span><span class="op">)</span></span>
<span><span class="va">n_gr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">x</span><span class="op">@</span><span class="va">n_total</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="va">n_gr</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span>, las <span class="op">=</span> <span class="fl">1</span>, xlab <span class="op">=</span> <span class="st">"Number"</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"Number of gr in background"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/text.html" class="external-link">text</a></span><span class="op">(</span><span class="va">n_gr</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="va">n_gr</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_use_background_files/figure-html/unnamed-chunk-11-1.png" width="960" style="display: block; margin: auto;"></p>
<p>It is quite astonishing that there are only 62 TFBS peaks in heterochromatin regions, but the heterochromatin regions cover more than 90% of the genome. We check the top most significant GO terms:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tb</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable</a></span><span class="op">(</span><span class="va">res_list</span><span class="op">[[</span><span class="st">"Heterochrom/lo"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##           id                        description genome_fraction</span></span>
<span><span class="co">## 1 GO:0032940                  secretion by cell      0.08585423</span></span>
<span><span class="co">## 2 GO:1903530    regulation of secretion by cell      0.06286739</span></span>
<span><span class="co">## 3 GO:0140352                   export from cell      0.09121197</span></span>
<span><span class="co">## 4 GO:0051046            regulation of secretion      0.06874861</span></span>
<span><span class="co">## 5 GO:0046903                          secretion      0.09709812</span></span>
<span><span class="co">## 6 GO:0032880 regulation of protein localization      0.08021046</span></span>
<span><span class="co">##   observed_region_hits fold_enrichment      p_value    p_adjust mean_tss_dist</span></span>
<span><span class="co">## 1                   18        3.381576 2.908187e-06 0.001418865        297080</span></span>
<span><span class="co">## 2                   15        3.848346 5.165993e-06 0.001418865        292459</span></span>
<span><span class="co">## 3                   18        3.182944 6.799673e-06 0.001418865        297080</span></span>
<span><span class="co">## 4                   15        3.519133 1.505106e-05 0.002012737        292459</span></span>
<span><span class="co">## 5                   18        2.989992 1.607618e-05 0.002012737        297080</span></span>
<span><span class="co">## 6                   16        3.217343 2.228778e-05 0.002325358        394119</span></span>
<span><span class="co">##   observed_gene_hits gene_set_size fold_enrichment_hyper p_value_hyper</span></span>
<span><span class="co">## 1                 15           715              3.669960  1.211072e-05</span></span>
<span><span class="co">## 2                 12           494              4.249428  2.388200e-05</span></span>
<span><span class="co">## 3                 15           774              3.390209  3.063420e-05</span></span>
<span><span class="co">## 4                 12           550              3.816759  6.766322e-05</span></span>
<span><span class="co">## 5                 15           837              3.135032  7.478033e-05</span></span>
<span><span class="co">## 6                 12           731              2.871706  9.100576e-04</span></span>
<span><span class="co">##   p_adjust_hyper</span></span>
<span><span class="co">## 1    0.001895328</span></span>
<span><span class="co">## 2    0.002491689</span></span>
<span><span class="co">## 3    0.002576789</span></span>
<span><span class="co">## 4    0.002976669</span></span>
<span><span class="co">## 5    0.002976669</span></span>
<span><span class="co">## 6    0.008901501</span></span></code></pre>
<p>Take the first term “<a href="GO:0032940" class="uri">GO:0032940</a>” for example, we can actually see, although there are only 62 TFBS peaks, 20 of them are associated with “<a href="GO:0032940" class="uri">GO:0032940</a>” and it is more than five times higher than the expected number: 62*0.0844 = 5.2.</p>
<p>However, the mean distance of “<a href="GO:0032940" class="uri">GO:0032940</a>” associated TFBS peaks to TSS in the heterochormatin background is 290kb, and mean distance to TSS of all peaks in heterochromatin is around 350kb, which is very far from TSS. With larger distance to TSS, the more we need to be careful with the relaiblity of the associations.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">8</span>, <span class="fl">4</span>, <span class="fl">1</span><span class="op">)</span>, xpd <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span></span>
<span><span class="va">dist_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">res_list</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">gra</span> <span class="op">=</span> <span class="fu"><a href="../reference/getRegionGeneAssociations-dispatch.html">getRegionGeneAssociations</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">gra</span><span class="op">$</span><span class="va">dist_to_TSS</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html" class="external-link">barplot</a></span><span class="op">(</span><span class="va">dist_list</span>, horiz <span class="op">=</span> <span class="cn">TRUE</span>, las <span class="op">=</span> <span class="fl">1</span>, xlab <span class="op">=</span> <span class="st">"mean distance to TSS"</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"mean distance to TSS"</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_use_background_files/figure-html/unnamed-chunk-13-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Therefore, we will not consider the category of heterochromatin in the analysis because although it generates many significant terms statistically, whether they make biological sense needs to be questioned.</p>
<p>In the end, we only compare the significant GO terms from TFBS peaks by taking promoters and enhancers as backgrounds. We use a loose cutoff for adjusted p-values (0.1) because more GO terms will give a better clustering.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig_lt2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">res_list</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tb</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span>    <span class="va">tb</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">tb</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/simplifyEnrichment/man/simplifyGOFromMultipleLists.html" class="external-link">simplifyGOFromMultipleLists</a></span><span class="op">(</span><span class="va">sig_lt2</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_use_background_files/figure-html/unnamed-chunk-14-1.png" width="960" style="display: block; margin: auto;"></p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">## Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">## Running under: macOS Ventura 13.2.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] C/UTF-8/C/C/C/C</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: Europe/Berlin</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] grid      stats4    stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] simplifyEnrichment_1.11.1 eulerr_7.0.0             </span></span>
<span><span class="co">## [3] rGREAT_2.5.4              GenomicRanges_1.52.1     </span></span>
<span><span class="co">## [5] GenomeInfoDb_1.36.4       IRanges_2.36.0           </span></span>
<span><span class="co">## [7] S4Vectors_0.40.2          BiocGenerics_0.48.1      </span></span>
<span><span class="co">## [9] knitr_1.44               </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3                      </span></span>
<span><span class="co">##   [2] jsonlite_1.8.8                          </span></span>
<span><span class="co">##   [3] shape_1.4.6                             </span></span>
<span><span class="co">##   [4] magrittr_2.0.3                          </span></span>
<span><span class="co">##   [5] magick_2.8.0                            </span></span>
<span><span class="co">##   [6] GenomicFeatures_1.52.2                  </span></span>
<span><span class="co">##   [7] rmarkdown_2.25                          </span></span>
<span><span class="co">##   [8] GlobalOptions_0.1.2                     </span></span>
<span><span class="co">##   [9] fs_1.6.3                                </span></span>
<span><span class="co">##  [10] BiocIO_1.10.0                           </span></span>
<span><span class="co">##  [11] zlibbioc_1.46.0                         </span></span>
<span><span class="co">##  [12] ragg_1.2.6                              </span></span>
<span><span class="co">##  [13] vctrs_0.6.4                             </span></span>
<span><span class="co">##  [14] Cairo_1.6-2                             </span></span>
<span><span class="co">##  [15] memoise_2.0.1                           </span></span>
<span><span class="co">##  [16] Rsamtools_2.16.0                        </span></span>
<span><span class="co">##  [17] RCurl_1.98-1.12                         </span></span>
<span><span class="co">##  [18] htmltools_0.5.7                         </span></span>
<span><span class="co">##  [19] S4Arrays_1.0.6                          </span></span>
<span><span class="co">##  [20] TxDb.Hsapiens.UCSC.hg19.knownGene_3.2.2 </span></span>
<span><span class="co">##  [21] progress_1.2.2                          </span></span>
<span><span class="co">##  [22] curl_5.1.0                              </span></span>
<span><span class="co">##  [23] sass_0.4.8                              </span></span>
<span><span class="co">##  [24] bslib_0.6.1                             </span></span>
<span><span class="co">##  [25] htmlwidgets_1.6.2                       </span></span>
<span><span class="co">##  [26] desc_1.4.2                              </span></span>
<span><span class="co">##  [27] cachem_1.0.8                            </span></span>
<span><span class="co">##  [28] GenomicAlignments_1.36.0                </span></span>
<span><span class="co">##  [29] mime_0.12                               </span></span>
<span><span class="co">##  [30] lifecycle_1.0.4                         </span></span>
<span><span class="co">##  [31] iterators_1.0.14                        </span></span>
<span><span class="co">##  [32] pkgconfig_2.0.3                         </span></span>
<span><span class="co">##  [33] Matrix_1.6-1.1                          </span></span>
<span><span class="co">##  [34] R6_2.5.1                                </span></span>
<span><span class="co">##  [35] fastmap_1.1.1                           </span></span>
<span><span class="co">##  [36] GenomeInfoDbData_1.2.10                 </span></span>
<span><span class="co">##  [37] MatrixGenerics_1.12.3                   </span></span>
<span><span class="co">##  [38] shiny_1.8.0                             </span></span>
<span><span class="co">##  [39] clue_0.3-65                             </span></span>
<span><span class="co">##  [40] digest_0.6.33                           </span></span>
<span><span class="co">##  [41] colorspace_2.1-0                        </span></span>
<span><span class="co">##  [42] AnnotationDbi_1.62.2                    </span></span>
<span><span class="co">##  [43] rprojroot_2.0.3                         </span></span>
<span><span class="co">##  [44] textshaping_0.3.7                       </span></span>
<span><span class="co">##  [45] RSQLite_2.3.1                           </span></span>
<span><span class="co">##  [46] org.Hs.eg.db_3.17.0                     </span></span>
<span><span class="co">##  [47] filelock_1.0.2                          </span></span>
<span><span class="co">##  [48] fansi_1.0.5                             </span></span>
<span><span class="co">##  [49] httr_1.4.7                              </span></span>
<span><span class="co">##  [50] polyclip_1.10-6                         </span></span>
<span><span class="co">##  [51] abind_1.4-5                             </span></span>
<span><span class="co">##  [52] compiler_4.3.1                          </span></span>
<span><span class="co">##  [53] bit64_4.0.5                             </span></span>
<span><span class="co">##  [54] doParallel_1.0.17                       </span></span>
<span><span class="co">##  [55] BiocParallel_1.34.2                     </span></span>
<span><span class="co">##  [56] DBI_1.1.3                               </span></span>
<span><span class="co">##  [57] biomaRt_2.56.1                          </span></span>
<span><span class="co">##  [58] rappdirs_0.3.3                          </span></span>
<span><span class="co">##  [59] proxyC_0.3.3                            </span></span>
<span><span class="co">##  [60] DelayedArray_0.26.7                     </span></span>
<span><span class="co">##  [61] rjson_0.2.21                            </span></span>
<span><span class="co">##  [62] tools_4.3.1                             </span></span>
<span><span class="co">##  [63] httpuv_1.6.13                           </span></span>
<span><span class="co">##  [64] glue_1.6.2                              </span></span>
<span><span class="co">##  [65] restfulr_0.0.15                         </span></span>
<span><span class="co">##  [66] GOSemSim_2.26.1                         </span></span>
<span><span class="co">##  [67] promises_1.2.1                          </span></span>
<span><span class="co">##  [68] polylabelr_0.2.0                        </span></span>
<span><span class="co">##  [69] cluster_2.1.4                           </span></span>
<span><span class="co">##  [70] generics_0.1.3                          </span></span>
<span><span class="co">##  [71] hms_1.1.3                               </span></span>
<span><span class="co">##  [72] xml2_1.3.6                              </span></span>
<span><span class="co">##  [73] utf8_1.2.3                              </span></span>
<span><span class="co">##  [74] XVector_0.40.0                          </span></span>
<span><span class="co">##  [75] foreach_1.5.2                           </span></span>
<span><span class="co">##  [76] pillar_1.9.0                            </span></span>
<span><span class="co">##  [77] stringr_1.5.0                           </span></span>
<span><span class="co">##  [78] later_1.3.2                             </span></span>
<span><span class="co">##  [79] circlize_0.4.15                         </span></span>
<span><span class="co">##  [80] dplyr_1.1.3                             </span></span>
<span><span class="co">##  [81] BiocFileCache_2.8.0                     </span></span>
<span><span class="co">##  [82] lattice_0.21-9                          </span></span>
<span><span class="co">##  [83] rtracklayer_1.60.1                      </span></span>
<span><span class="co">##  [84] bit_4.0.5                               </span></span>
<span><span class="co">##  [85] tidyselect_1.2.0                        </span></span>
<span><span class="co">##  [86] GO.db_3.17.0                            </span></span>
<span><span class="co">##  [87] ComplexHeatmap_2.18.0                   </span></span>
<span><span class="co">##  [88] tm_0.7-11                               </span></span>
<span><span class="co">##  [89] Biostrings_2.68.1                       </span></span>
<span><span class="co">##  [90] NLP_0.2-1                               </span></span>
<span><span class="co">##  [91] SummarizedExperiment_1.30.2             </span></span>
<span><span class="co">##  [92] xfun_0.40                               </span></span>
<span><span class="co">##  [93] Biobase_2.60.0                          </span></span>
<span><span class="co">##  [94] matrixStats_1.2.0                       </span></span>
<span><span class="co">##  [95] DT_0.30                                 </span></span>
<span><span class="co">##  [96] stringi_1.7.12                          </span></span>
<span><span class="co">##  [97] yaml_2.3.7                              </span></span>
<span><span class="co">##  [98] TxDb.Hsapiens.UCSC.hg38.knownGene_3.17.0</span></span>
<span><span class="co">##  [99] evaluate_0.22                           </span></span>
<span><span class="co">## [100] codetools_0.2-19                        </span></span>
<span><span class="co">## [101] tibble_3.2.1                            </span></span>
<span><span class="co">## [102] cli_3.6.2                               </span></span>
<span><span class="co">## [103] RcppParallel_5.1.7                      </span></span>
<span><span class="co">## [104] xtable_1.8-4                            </span></span>
<span><span class="co">## [105] systemfonts_1.0.5                       </span></span>
<span><span class="co">## [106] jquerylib_0.1.4                         </span></span>
<span><span class="co">## [107] Rcpp_1.0.11                             </span></span>
<span><span class="co">## [108] dbplyr_2.3.4                            </span></span>
<span><span class="co">## [109] png_0.1-8                               </span></span>
<span><span class="co">## [110] XML_3.99-0.14                           </span></span>
<span><span class="co">## [111] parallel_4.3.1                          </span></span>
<span><span class="co">## [112] ellipsis_0.3.2                          </span></span>
<span><span class="co">## [113] pkgdown_2.0.7                           </span></span>
<span><span class="co">## [114] blob_1.2.4                              </span></span>
<span><span class="co">## [115] prettyunits_1.2.0                       </span></span>
<span><span class="co">## [116] bitops_1.0-7                            </span></span>
<span><span class="co">## [117] slam_0.1-50                             </span></span>
<span><span class="co">## [118] purrr_1.0.2                             </span></span>
<span><span class="co">## [119] crayon_1.5.2                            </span></span>
<span><span class="co">## [120] GetoptLong_1.0.5                        </span></span>
<span><span class="co">## [121] rlang_1.1.2                             </span></span>
<span><span class="co">## [122] KEGGREST_1.40.1</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
