<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Suppl 5. Compare different genome versions • rGREAT</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Suppl 5. Compare different genome versions">
<meta property="og:description" content="rGREAT">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rGREAT</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.5.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Online GREAT</li>
    <li>
      <a href="../articles/online-GREAT.html">Analyze with online GREAT</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Local GREAT</li>
    <li>
      <a href="../articles/local-GREAT.html">Analyze with local GREAT</a>
    </li>
    <li>
      <a href="../articles/other-geneset-databases.html">Work with other geneset databases</a>
    </li>
    <li>
      <a href="../articles/other-organisms.html">Work with other organisms</a>
    </li>
    <li>
      <a href="../articles/group_associations.html">Get gene-region associations for a group of GO terms</a>
    </li>
    <li class="divider">
    </li>
<li class="dropdown-header">Supplementaries in the rGREAT paper</li>
    <li>
      <a href="../articles/suppl_use_background.html">Suppl 1. Use background regions</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_GO.html">Suppl 2. Compare GO annotations from online and local GREAT</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_online_and_local.html">Suppl 3. Compare online and local GREAT</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_tss.html">Suppl 4. Compare different TSS sources</a>
    </li>
    <li>
      <a href="../articles/suppl_compare_genome_version.html">Suppl 5. Compare different genome versions</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/jokergoo/rGREAT/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Suppl 5. Compare different genome versions</h1>
                        <h4 data-toc-skip class="author">Zuguang Gu ( <a href="mailto:z.gu@dkfz.de" class="email">z.gu@dkfz.de</a> )</h4>
            
            <h4 data-toc-skip class="date">2024-02-27</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/jokergoo/rGREAT/blob/HEAD/vignettes/suppl_compare_genome_versions.Rmd" class="external-link"><code>vignettes/suppl_compare_genome_versions.Rmd</code></a></small>
      <div class="hidden name"><code>suppl_compare_genome_versions.Rmd</code></div>

    </div>

    
    
<p>In this document, we demonstrate the consequence of using a wrong genome version in the GREAT analysis.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/rGREAT" class="external-link">rGREAT</a></span><span class="op">)</span></span></code></pre></div>
<p>We use a dataset with hg38 genome from <a href="https://genome.ucsc.edu/cgi-bin/hgTables" class="external-link">UCSC table browser</a>. The parameters are as follows:</p>
<pre><code>clade = Mammal
genome = Human
assembly = GRCh38/hg38
group = Regulation
track = TF ChIP
table = A549 MYC (encTfChipPkENCFF542GMN)</code></pre>
<p>We perform local GREAT with hg38 genome as well as hg19 genome.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="st">"data/A549_MYC_encTfChipPkENCFF542GMN_hg38.bed"</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">=</span> <span class="va">df</span><span class="op">[</span><span class="va">df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span>, <span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">gr</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="va">df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, ranges <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">+</span><span class="fl">1</span>, <span class="va">df</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">res_hg19</span> <span class="op">=</span> <span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"TxDb.Hsapiens.UCSC.hg19.knownGene"</span>, min_gene_set_size <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Inconsistent coordinate on 'chr5': Max end position (181261577) is</span></span>
<span><span class="co">## larger than chromosome length (180915260).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Inconsistent coordinate on 'chr17': Max end position (83110490) is</span></span>
<span><span class="co">## larger than chromosome length (81195210).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Inconsistent coordinate on 'chr18': Max end position (80034737) is</span></span>
<span><span class="co">## larger than chromosome length (78077248).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Inconsistent coordinate on 'chr20': Max end position (64131770) is</span></span>
<span><span class="co">## larger than chromosome length (63025520).</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Inconsistent coordinate on 'chrX': Max end position (155881472) is</span></span>
<span><span class="co">## larger than chromosome length (155270560).</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_hg38</span> <span class="op">=</span> <span class="fu"><a href="../reference/great.html">great</a></span><span class="op">(</span><span class="va">gr</span>, <span class="st">"GO:BP"</span>, <span class="st">"TxDb.Hsapiens.UCSC.hg38.knownGene"</span>, min_gene_set_size <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="va">tb_hg19</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable</a></span><span class="op">(</span><span class="va">res_hg19</span><span class="op">)</span></span>
<span><span class="va">tb_hg38</span> <span class="op">=</span> <span class="fu"><a href="../reference/getEnrichmentTable-dispatch.html">getEnrichmentTable</a></span><span class="op">(</span><span class="va">res_hg38</span><span class="op">)</span></span></code></pre></div>
<p>You can already see there are warnings if using the wrong genome version.</p>
<p>The following plot shows overlap of significant GO terms. There are quite big disagreements.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jolars/eulerr" class="external-link">eulerr</a></span><span class="op">)</span></span>
<span><span class="va">lt</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    hg19 <span class="op">=</span> <span class="va">tb_hg19</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">tb_hg19</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">]</span>,</span>
<span>    hg38 <span class="op">=</span> <span class="va">tb_hg38</span><span class="op">$</span><span class="va">id</span><span class="op">[</span><span class="va">tb_hg38</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://jolars.github.io/eulerr/reference/euler.html" class="external-link">euler</a></span><span class="op">(</span><span class="va">lt</span><span class="op">)</span>, quantities <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_compare_genome_versions_files/figure-html/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;"></p>
<p>To directly compare the two results, next we only use the common GO terms in the two results, and make the two result tables having the same row orders.</p>
<p>In the following code, we also randomly permute <code>cn</code> to remove some pre-orderings in the result tables, which may affect later visualizations.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tb_hg19</span><span class="op">)</span> <span class="op">=</span> <span class="va">tb_hg19</span><span class="op">$</span><span class="va">id</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tb_hg38</span><span class="op">)</span> <span class="op">=</span> <span class="va">tb_hg38</span><span class="op">$</span><span class="va">id</span></span>
<span><span class="va">cn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">intersect</a></span><span class="op">(</span><span class="va">tb_hg19</span><span class="op">$</span><span class="va">id</span>, <span class="va">tb_hg38</span><span class="op">$</span><span class="va">id</span><span class="op">)</span></span>
<span><span class="va">cn</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">cn</span><span class="op">)</span></span>
<span><span class="va">tb_hg19</span> <span class="op">=</span> <span class="va">tb_hg19</span><span class="op">[</span><span class="va">cn</span>, <span class="op">]</span></span>
<span><span class="va">tb_hg38</span> <span class="op">=</span> <span class="va">tb_hg38</span><span class="op">[</span><span class="va">cn</span>, <span class="op">]</span></span></code></pre></div>
<p>We first make a scatter plot of log2 fold enrichment (<span class="math inline">\(log_2(obs/exp)\)</span>) in the two results. The plot shows the relation is not very linear.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">tb_hg19</span><span class="op">$</span><span class="va">fold_enrichment</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">tb_hg38</span><span class="op">$</span><span class="va">fold_enrichment</span><span class="op">)</span>, pch <span class="op">=</span> <span class="fl">16</span>, </span>
<span>    col <span class="op">=</span> <span class="st">"#00000020"</span>, xlab <span class="op">=</span> <span class="st">"hg19"</span>, ylab <span class="op">=</span> <span class="st">"hg38"</span>, main <span class="op">=</span> <span class="st">"log2 fold change"</span>,</span>
<span>    xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">6</span>, <span class="fl">6</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">6</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_compare_genome_versions_files/figure-html/unnamed-chunk-6-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Next we compare the genome fraction of the genomic domain associated to each GO term. The plot shows the relation is very linear. It makes sense because in the different genome versions, genes are more like to be shifted while the widths of corresponding genomic domains are not affacted too much.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tb_hg19</span><span class="op">$</span><span class="va">genome_fraction</span>, <span class="va">tb_hg38</span><span class="op">$</span><span class="va">genome_fraction</span>, pch <span class="op">=</span> <span class="fl">16</span>, log <span class="op">=</span> <span class="st">"xy"</span>,</span>
<span>    col <span class="op">=</span> <span class="st">"#00000020"</span>, xlab <span class="op">=</span> <span class="st">"hg19"</span>, ylab <span class="op">=</span> <span class="st">"hg38"</span>, main <span class="op">=</span> <span class="st">"genome_fraction"</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_compare_genome_versions_files/figure-html/unnamed-chunk-7-1.png" width="672" style="display: block; margin: auto;"></p>
<p>The number of input regions that fall into the genomic domain associated with a certain GO term is denoted as a random variable <span class="math inline">\(\textit{X}\)</span>. <span class="math inline">\(\textit{X}\)</span> follows a Binomial distribution: <span class="math inline">\(\textit{X} \sim B(\textit{p}, \textit{N})\)</span>, where <span class="math inline">\(\textit{N}\)</span> is the total number of input regions and <span class="math inline">\(\textit{p}\)</span> is the genomic fraction of a certain functional genomic domain. The <em>p</em>-value is calculated as <span class="math inline">\(\textit{P} = Pr(\textit{X} \geq \textit{n})\)</span>. As we have demonstrated in previous plots, <span class="math inline">\(\textit{N}\)</span> is almost unchanged, and <span class="math inline">\(\textit{p}\)</span> is highly similar in the two genome versions. The only thing that affects <em>p</em>-values is <span class="math inline">\(\textit{n}\)</span>, i.e. the number of input regions that fall into the genomic domain.</p>
<p>Next we compare the number of regions in a certain genomic domain (the column <code>observed_region_hits</code> in the result table) in the two genome versions. We also highlight whether the GO term is significant in both genome versions or only in one genome version.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l1</span> <span class="op">=</span> <span class="va">tb_hg19</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">tb_hg38</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&gt;</span> <span class="fl">0.01</span>   <span class="co"># only in hg19</span></span>
<span><span class="va">l2</span> <span class="op">=</span> <span class="va">tb_hg19</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&gt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">tb_hg38</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.01</span>   <span class="co"># only in hg38</span></span>
<span><span class="va">l3</span> <span class="op">=</span> <span class="va">tb_hg19</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">tb_hg38</span><span class="op">$</span><span class="va">p_adjust</span> <span class="op">&lt;</span> <span class="fl">0.01</span>   <span class="co"># both</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tb_hg19</span><span class="op">$</span><span class="va">observed_region_hits</span>, <span class="va">tb_hg38</span><span class="op">$</span><span class="va">observed_region_hits</span>, pch <span class="op">=</span> <span class="fl">16</span>, log <span class="op">=</span> <span class="st">"xy"</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">l1</span>, <span class="st">"#FF000040"</span>, <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">l2</span>, <span class="st">"#00FF0040"</span>, <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">l3</span>, <span class="st">"#0000FF40"</span>, <span class="st">"#00000060"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>, </span>
<span>    cex <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">l1</span> <span class="op">|</span> <span class="va">l2</span> <span class="op">|</span> <span class="va">l3</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"hg19"</span>, ylab <span class="op">=</span> <span class="st">"hg38"</span>, main <span class="op">=</span> <span class="st">"observed_region_hits"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>, pch <span class="op">=</span> <span class="fl">16</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span>, <span class="st">"green"</span><span class="op">)</span>, </span>
<span>    legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"only sig in hg19"</span>, <span class="st">"sig in both"</span>, <span class="st">"only sig in hg38"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_compare_genome_versions_files/figure-html/unnamed-chunk-8-1.png" width="672" style="display: block; margin: auto;"></p>
<p>Based on the plot above, we can see using the wrong genome versions mainly affects these functional genomic domains where only a few input regions are in. Using different genome versions, input regions may be in a genomic domain or not, which especially increases the (relative) variability of GO terms with small “observed_region_hits” (bottom left part in the plot).</p>
<p>We can validate it by comparing the average distance to gene TSS in each functional domain. In the following plot, we can easily see when a GO term has a smaller “observed_region_hits” (red dots), the average distances from input regions to the associated TSSs are more variable between the two genome versions.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">l</span> <span class="op">=</span> <span class="va">tb_hg38</span><span class="op">$</span><span class="va">observed_region_hits</span> <span class="op">&lt;</span> <span class="fl">200</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tb_hg19</span><span class="op">$</span><span class="va">mean_tss_dist</span>, <span class="va">tb_hg38</span><span class="op">$</span><span class="va">mean_tss_dist</span>, pch <span class="op">=</span> <span class="fl">16</span>, log <span class="op">=</span> <span class="st">"xy"</span>,</span>
<span>    col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">l</span>, <span class="st">"#FF000020"</span>, <span class="st">"#0000FF60"</span><span class="op">)</span>, xlim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e2</span>, <span class="fl">1e6</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1e2</span>, <span class="fl">1e6</span><span class="op">)</span>,</span>
<span>    xlab <span class="op">=</span> <span class="st">"hg19"</span>, ylab <span class="op">=</span> <span class="st">"hg38"</span>, main <span class="op">=</span> <span class="st">"mean_tss_dist"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/legend.html" class="external-link">legend</a></span><span class="op">(</span><span class="st">"topleft"</span>, pch <span class="op">=</span> <span class="fl">16</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"blue"</span><span class="op">)</span>,</span>
<span>    legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"small n_obs in hg38 (&lt; 200)"</span>, <span class="st">"large n_obs in hg38 (&gt;= 200)"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="suppl_compare_genome_versions_files/figure-html/unnamed-chunk-9-1.png" width="672" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>This analysis shows using the wrong genome version greatly affects the enrichment results, especially for the functional terms which only have a few associated input regions.</p>
<p><br><br><br><br><br></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Zuguang Gu.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  <style>nav[data-toggle='toc'] .nav .nav {display: block;}</style>
</body>
</html>
